{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "RDKit WARNING: [15:57:24] Enabling RDKit 2019.09.2 jupyter extensions\n"
     ]
    }
   ],
   "source": [
    "import sklearn as sk\n",
    "import rdkit as rd\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit.Chem import Draw\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package sklearn:\n",
      "\n",
      "NAME\n",
      "    sklearn\n",
      "\n",
      "DESCRIPTION\n",
      "    Machine learning module for Python\n",
      "    ==================================\n",
      "    \n",
      "    sklearn is a Python module integrating classical machine\n",
      "    learning algorithms in the tightly-knit world of scientific Python\n",
      "    packages (numpy, scipy, matplotlib).\n",
      "    \n",
      "    It aims to provide simple and efficient solutions to learning problems\n",
      "    that are accessible to everybody and reusable in various contexts:\n",
      "    machine-learning as a versatile tool for science and engineering.\n",
      "    \n",
      "    See http://scikit-learn.org for complete documentation.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    __check_build (package)\n",
      "    _build_utils (package)\n",
      "    _config\n",
      "    _isotonic\n",
      "    base\n",
      "    calibration\n",
      "    cluster (package)\n",
      "    compose (package)\n",
      "    conftest\n",
      "    covariance (package)\n",
      "    cross_decomposition (package)\n",
      "    datasets (package)\n",
      "    decomposition (package)\n",
      "    discriminant_analysis\n",
      "    dummy\n",
      "    ensemble (package)\n",
      "    exceptions\n",
      "    experimental (package)\n",
      "    externals (package)\n",
      "    feature_extraction (package)\n",
      "    feature_selection (package)\n",
      "    gaussian_process (package)\n",
      "    impute (package)\n",
      "    inspection (package)\n",
      "    isotonic\n",
      "    kernel_approximation\n",
      "    kernel_ridge\n",
      "    linear_model (package)\n",
      "    manifold (package)\n",
      "    metrics (package)\n",
      "    mixture (package)\n",
      "    model_selection (package)\n",
      "    multiclass\n",
      "    multioutput\n",
      "    naive_bayes\n",
      "    neighbors (package)\n",
      "    neural_network (package)\n",
      "    pipeline\n",
      "    preprocessing (package)\n",
      "    random_projection\n",
      "    semi_supervised (package)\n",
      "    setup\n",
      "    svm (package)\n",
      "    tests (package)\n",
      "    tree (package)\n",
      "    utils (package)\n",
      "\n",
      "FUNCTIONS\n",
      "    clone(estimator, safe=True)\n",
      "        Constructs a new estimator with the same parameters.\n",
      "        \n",
      "        Clone does a deep copy of the model in an estimator\n",
      "        without actually copying attached data. It yields a new estimator\n",
      "        with the same parameters that has not been fit on any data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        estimator : estimator object, or list, tuple or set of objects\n",
      "            The estimator or group of estimators to be cloned\n",
      "        \n",
      "        safe : boolean, optional\n",
      "            If safe is false, clone will fall back to a deep copy on objects\n",
      "            that are not estimators.\n",
      "    \n",
      "    config_context(**new_config)\n",
      "        Context manager for global scikit-learn configuration\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        assume_finite : bool, optional\n",
      "            If True, validation for finiteness will be skipped,\n",
      "            saving time, but leading to potential crashes. If\n",
      "            False, validation for finiteness will be performed,\n",
      "            avoiding error.  Global default: False.\n",
      "        \n",
      "        working_memory : int, optional\n",
      "            If set, scikit-learn will attempt to limit the size of temporary arrays\n",
      "            to this number of MiB (per job when parallelised), often saving both\n",
      "            computation time and memory on expensive operations that can be\n",
      "            performed in chunks. Global default: 1024.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        All settings, not just those presently modified, will be returned to\n",
      "        their previous values when the context manager is exited. This is not\n",
      "        thread-safe.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import sklearn\n",
      "        >>> from sklearn.utils.validation import assert_all_finite\n",
      "        >>> with sklearn.config_context(assume_finite=True):\n",
      "        ...     assert_all_finite([float('nan')])\n",
      "        >>> with sklearn.config_context(assume_finite=True):\n",
      "        ...     with sklearn.config_context(assume_finite=False):\n",
      "        ...         assert_all_finite([float('nan')])\n",
      "        ... # doctest: +ELLIPSIS\n",
      "        Traceback (most recent call last):\n",
      "        ...\n",
      "        ValueError: Input contains NaN, ...\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        set_config: Set global scikit-learn configuration\n",
      "        get_config: Retrieve current values of the global configuration\n",
      "    \n",
      "    get_config()\n",
      "        Retrieve current values for configuration set by :func:`set_config`\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        config : dict\n",
      "            Keys are parameter names that can be passed to :func:`set_config`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        config_context: Context manager for global scikit-learn configuration\n",
      "        set_config: Set global scikit-learn configuration\n",
      "    \n",
      "    set_config(assume_finite=None, working_memory=None, print_changed_only=None)\n",
      "        Set global scikit-learn configuration\n",
      "        \n",
      "        .. versionadded:: 0.19\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        assume_finite : bool, optional\n",
      "            If True, validation for finiteness will be skipped,\n",
      "            saving time, but leading to potential crashes. If\n",
      "            False, validation for finiteness will be performed,\n",
      "            avoiding error.  Global default: False.\n",
      "        \n",
      "            .. versionadded:: 0.19\n",
      "        \n",
      "        working_memory : int, optional\n",
      "            If set, scikit-learn will attempt to limit the size of temporary arrays\n",
      "            to this number of MiB (per job when parallelised), often saving both\n",
      "            computation time and memory on expensive operations that can be\n",
      "            performed in chunks. Global default: 1024.\n",
      "        \n",
      "            .. versionadded:: 0.20\n",
      "        \n",
      "        print_changed_only : bool, optional\n",
      "            If True, only the parameters that were set to non-default\n",
      "            values will be printed when printing an estimator. For example,\n",
      "            ``print(SVC())`` while True will only print 'SVC()' while the default\n",
      "            behaviour would be to print 'SVC(C=1.0, cache_size=200, ...)' with\n",
      "            all the non-changed parameters.\n",
      "        \n",
      "            .. versionadded:: 0.21\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        config_context: Context manager for global scikit-learn configuration\n",
      "        get_config: Retrieve current values of the global configuration\n",
      "    \n",
      "    show_versions()\n",
      "        Print useful debugging information\n",
      "\n",
      "DATA\n",
      "    __SKLEARN_SETUP__ = False\n",
      "    __all__ = ['calibration', 'cluster', 'covariance', 'cross_decompositio...\n",
      "\n",
      "VERSION\n",
      "    0.21.3\n",
      "\n",
      "FILE\n",
      "    /opt/anaconda3/envs/moldist2/lib/python3.7/site-packages/sklearn/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(sk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package rdkit.Chem in rdkit:\n",
      "\n",
      "NAME\n",
      "    rdkit.Chem - A module for molecules and stuff\n",
      "\n",
      "DESCRIPTION\n",
      "    see Chem/index.html in the doc tree for documentation\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    AllChem\n",
      "    AtomPairs (package)\n",
      "    BRICS\n",
      "    BuildFragmentCatalog\n",
      "    ChemUtils (package)\n",
      "    ChemicalFeatures\n",
      "    ChemicalForceFields\n",
      "    Crippen\n",
      "    DSViewer\n",
      "    Descriptors\n",
      "    Descriptors3D\n",
      "    Draw (package)\n",
      "    EState (package)\n",
      "    EnumerateHeterocycles\n",
      "    EnumerateStereoisomers\n",
      "    FastSDMolSupplier\n",
      "    FeatFinderCLI\n",
      "    FeatMaps (package)\n",
      "    Features (package)\n",
      "    FilterCatalog\n",
      "    Fingerprints (package)\n",
      "    Fraggle (package)\n",
      "    FragmentCatalog\n",
      "    FragmentMatcher\n",
      "    Fragments\n",
      "    FunctionalGroups\n",
      "    GraphDescriptors\n",
      "    Graphs\n",
      "    Lipinski\n",
      "    MACCSkeys\n",
      "    MCS\n",
      "    MolCatalog\n",
      "    MolDb (package)\n",
      "    MolKey (package)\n",
      "    MolStandardize (package)\n",
      "    MolSurf\n",
      "    PandasTools\n",
      "    Pharm2D (package)\n",
      "    Pharm3D (package)\n",
      "    PropertyMol\n",
      "    PyMol\n",
      "    QED\n",
      "    Randomize\n",
      "    Recap\n",
      "    ReducedGraphs\n",
      "    SATIS\n",
      "    SaltRemover\n",
      "    Scaffolds (package)\n",
      "    ShowMols\n",
      "    SimpleEnum (package)\n",
      "    Subshape (package)\n",
      "    Suppliers (package)\n",
      "    TemplateAlign\n",
      "    TorsionFingerprints\n",
      "    fmcs (package)\n",
      "    inchi\n",
      "    rdChemReactions\n",
      "    rdChemicalFeatures\n",
      "    rdCoordGen\n",
      "    rdDepictor\n",
      "    rdDistGeom\n",
      "    rdEHTTools\n",
      "    rdFMCS\n",
      "    rdFingerprintGenerator\n",
      "    rdForceFieldHelpers\n",
      "    rdFreeSASA\n",
      "    rdMMPA\n",
      "    rdMolAlign\n",
      "    rdMolCatalog\n",
      "    rdMolChemicalFeatures\n",
      "    rdMolDescriptors\n",
      "    rdMolHash\n",
      "    rdMolInterchange\n",
      "    rdMolTransforms\n",
      "    rdPartialCharges\n",
      "    rdRGroupDecomposition\n",
      "    rdReducedGraphs\n",
      "    rdShapeHelpers\n",
      "    rdStructChecker\n",
      "    rdSubstructLibrary\n",
      "    rdchem\n",
      "    rdfiltercatalog\n",
      "    rdfragcatalog\n",
      "    rdinchi\n",
      "    rdmolfiles\n",
      "    rdmolops\n",
      "    rdqueries\n",
      "    rdtrajectory\n",
      "\n",
      "FUNCTIONS\n",
      "    AddHs(...)\n",
      "        AddHs( (Mol)mol [, (bool)explicitOnly=False [, (bool)addCoords=False [, (AtomPairsParameters)onlyOnAtoms=None [, (bool)addResidueInfo=False]]]]) -> Mol :\n",
      "            Adds hydrogens to the graph of a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - explicitOnly: (optional) if this toggle is set, only explicit Hs will\n",
      "                  be added to the molecule.  Default value is 0 (add implicit and explicit Hs).\n",
      "            \n",
      "                - addCoords: (optional) if this toggle is set, The Hs will have 3D coordinates\n",
      "                  set.  Default value is 0 (no 3D coords).\n",
      "            \n",
      "                - onlyOnAtoms: (optional) if this sequence is provided, only these atoms will be\n",
      "                  considered to have Hs added to them\n",
      "            \n",
      "                - addResidueInfo: (optional) if this is true, add residue info to\n",
      "                  hydrogen atoms (useful for PDB files).\n",
      "            \n",
      "              RETURNS: a new molecule with added Hs\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "                - Much of the code assumes that Hs are not included in the molecular\n",
      "                  topology, so be *very* careful with the molecule that comes back from\n",
      "                  this function.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* AddHs(RDKit::ROMol [,bool=False [,bool=False [,boost::python::api::object=None [,bool=False]]]])\n",
      "    \n",
      "    AddRecursiveQuery(...)\n",
      "        AddRecursiveQuery( (Mol)mol, (Mol)query, (int)atomIdx [, (bool)preserveExistingQuery=True]) -> None :\n",
      "            Adds a recursive query to an atom\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - query: the molecule to be used as the recursive query (this will be copied)\n",
      "            \n",
      "                - atomIdx: the atom to modify\n",
      "            \n",
      "                - preserveExistingQuery: (optional) if this is set, existing query information on the atom will be preserved\n",
      "            \n",
      "              RETURNS: None\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AddRecursiveQuery(RDKit::ROMol {lvalue},RDKit::ROMol,unsigned int [,bool=True])\n",
      "    \n",
      "    AdjustQueryProperties(...)\n",
      "        AdjustQueryProperties( (Mol)mol [, (AtomPairsParameters)params=None]) -> Mol :\n",
      "            Returns a new molecule where the query properties of atoms have been modified.\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* AdjustQueryProperties(RDKit::ROMol [,boost::python::api::object=None])\n",
      "    \n",
      "    AssignAtomChiralTagsFromStructure(...)\n",
      "        AssignAtomChiralTagsFromStructure( (Mol)mol [, (int)confId=-1 [, (bool)replaceExistingTags=True]]) -> None :\n",
      "            Sets the chiral tags on a molecule's atoms based on \n",
      "              a 3D conformation.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - confId: the conformer id to use, -1 for the default \n",
      "                - replaceExistingTags: if True, existing stereochemistry information will be cleared \n",
      "                before running the calculation. \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignAtomChiralTagsFromStructure(RDKit::ROMol {lvalue} [,int=-1 [,bool=True]])\n",
      "    \n",
      "    AssignChiralTypesFromBondDirs(...)\n",
      "        AssignChiralTypesFromBondDirs( (Mol)mol [, (int)confId=-1 [, (bool)replaceExistingTags=True]]) -> None :\n",
      "            Uses bond directions to assign ChiralTypes to a molecule's atoms.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - confId: (optional) the conformation to use \n",
      "                - replaceExistingTags: (optional) replace any existing information about stereochemistry\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignChiralTypesFromBondDirs(RDKit::ROMol {lvalue} [,int=-1 [,bool=True]])\n",
      "    \n",
      "    AssignRadicals(...)\n",
      "        AssignRadicals( (Mol)mol) -> None :\n",
      "            Assigns radical counts to atoms\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignRadicals(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    AssignStereochemistry(...)\n",
      "        AssignStereochemistry( (Mol)mol [, (bool)cleanIt=False [, (bool)force=False [, (bool)flagPossibleStereoCenters=False]]]) -> None :\n",
      "            Does the CIP stereochemistry assignment \n",
      "              for the molecule's atoms (R/S) and double bond (Z/E).\n",
      "              Chiral atoms will have a property '_CIPCode' indicating\n",
      "              their chiral code.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - cleanIt: (optional) if provided, atoms with a chiral specifier that aren't\n",
      "                  actually chiral (e.g. atoms with duplicate substituents or only 2 substituents,\n",
      "                  etc.) will have their chiral code set to CHI_UNSPECIFIED. Bonds with \n",
      "                  STEREOCIS/STEREOTRANS specified that have duplicate substituents based upon the CIP \n",
      "                  atom ranks will be marked STEREONONE. \n",
      "                - force: (optional) causes the calculation to be repeated, even if it has already\n",
      "                  been done\n",
      "                - flagPossibleStereoCenters (optional)   set the _ChiralityPossible property on\n",
      "                  atoms that are possible stereocenters\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignStereochemistry(RDKit::ROMol {lvalue} [,bool=False [,bool=False [,bool=False]]])\n",
      "    \n",
      "    AssignStereochemistryFrom3D(...)\n",
      "        AssignStereochemistryFrom3D( (Mol)mol [, (int)confId=-1 [, (bool)replaceExistingTags=True]]) -> None :\n",
      "            Uses a conformer (should be 3D) to assign ChiralTypes to a molecule's atoms\n",
      "                    and stereo flags to its bonds\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - confId: (optional) the conformation to use \n",
      "                - replaceExistingTags: (optional) replace any existing information about stereochemistry\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void AssignStereochemistryFrom3D(RDKit::ROMol {lvalue} [,int=-1 [,bool=True]])\n",
      "    \n",
      "    AtomFromSmarts(...)\n",
      "        AtomFromSmarts( (str)SMARTS) -> Atom :\n",
      "            Construct an atom from a SMARTS string\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::Atom* AtomFromSmarts(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    AtomFromSmiles(...)\n",
      "        AtomFromSmiles( (str)SMILES) -> Atom :\n",
      "            Construct an atom from a SMILES string\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::Atom* AtomFromSmiles(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    BondFromSmarts(...)\n",
      "        BondFromSmarts( (str)SMILES) -> Bond :\n",
      "            Construct a bond from a SMARTS string\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::Bond* BondFromSmarts(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    BondFromSmiles(...)\n",
      "        BondFromSmiles( (str)SMILES) -> Bond :\n",
      "            Construct a bond from a SMILES string\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::Bond* BondFromSmiles(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    CanonSmiles(smi, useChiral=1)\n",
      "    \n",
      "    CanonicalRankAtoms(...)\n",
      "        CanonicalRankAtoms( (Mol)mol [, (bool)breakTies=True [, (bool)includeChirality=True [, (bool)includeIsotopes=True]]]) -> _vectj :\n",
      "            Returns the canonical atom ranking for each atom of a molecule fragment.\n",
      "              If breakTies is False, this returns the symmetry class for each atom.  The symmetry\n",
      "              class is used by the canonicalization routines to type each atom based on the whole\n",
      "              chemistry of the molecular graph.  Any atom with the same rank (symmetry class) is\n",
      "              indistinguishable.  For example:\n",
      "            \n",
      "                >>> mol = MolFromSmiles('C1NCN1')\n",
      "                >>> list(CanonicalRankAtoms(mol, breakTies=False))\n",
      "                [0,1,0,1]\n",
      "            \n",
      "              In this case the carbons have the same symmetry class and the nitrogens have the same\n",
      "              symmetry class.  From the perspective of the Molecular Graph, they are indentical.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - breakTies: (optional) force breaking of ranked ties [default=True]\n",
      "                - includeChirality: (optional) use chiral information when computing rank [default=True]\n",
      "                - includeIsotopes: (optional) use isotope information when computing rank [default=True]\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::vector<unsigned int, std::__1::allocator<unsigned int> > CanonicalRankAtoms(RDKit::ROMol [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    CanonicalRankAtomsInFragment(...)\n",
      "        CanonicalRankAtomsInFragment( (Mol)mol, (AtomPairsParameters)atomsToUse [, (AtomPairsParameters)bondsToUse=0 [, (AtomPairsParameters)atomSymbols=0 [, (bool)breakTies=True [, (bool)includeChirality=True [, (bool)includeIsotopes=True]]]]]) -> _vecti :\n",
      "            Returns the canonical atom ranking for each atom of a molecule fragment\n",
      "              See help(CanonicalRankAtoms) for more information.\n",
      "            \n",
      "               >>> mol = MolFromSmiles('C1NCN1.C1NCN1')\n",
      "               >>> list(CanonicalRankAtomsInFragment(mol, atomsToUse=range(0,4), breakTies=False))\n",
      "               [0,1,0,1,-1,-1,-1,-1]\n",
      "               >>> list(CanonicalRankAtomsInFragment(mol, atomsToUse=range(4,8), breakTies=False))\n",
      "               [-1,-1,-1,-1,0,1,0,1]\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - atomsToUse : a list of atoms to include in the fragment\n",
      "                - bondsToUse : (optional) a list of bonds to include in the fragment\n",
      "                  if not provided, all bonds between the atoms provided\n",
      "                  will be included.\n",
      "                - atomSymbols : (optional) a list with the symbols to use for the atoms\n",
      "                  in the SMILES. This should have be mol.GetNumAtoms() long.\n",
      "                - breakTies: (optional) force breaking of ranked ties\n",
      "                - includeChirality: (optional) use chiral information when computing rank [default=True]\n",
      "                - includeIsotopes: (optional) use isotope information when computing rank [default=True]\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::vector<int, std::__1::allocator<int> > CanonicalRankAtomsInFragment(RDKit::ROMol,boost::python::api::object [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=True [,bool=True [,bool=True]]]]])\n",
      "    \n",
      "    Cleanup(...)\n",
      "        Cleanup( (Mol)mol) -> None :\n",
      "            cleans up certain common bad functionalities in the molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void Cleanup(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    ClearMolSubstanceGroups(...)\n",
      "        ClearMolSubstanceGroups( (Mol)arg1) -> None :\n",
      "            removes all SubstanceGroups from a molecule (if any)\n",
      "        \n",
      "            C++ signature :\n",
      "                void ClearMolSubstanceGroups(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    CombineMols(...)\n",
      "        CombineMols( (Mol)mol1, (Mol)mol2 [, (Point3D)offset=<rdkit.Geometry.rdGeometry.Point3D object at 0x1180ff6c0>]) -> Mol :\n",
      "            Combine the atoms from two molecules to produce a third\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* CombineMols(RDKit::ROMol,RDKit::ROMol [,RDGeom::Point3D=<rdkit.Geometry.rdGeometry.Point3D object at 0x1180ff6c0>])\n",
      "    \n",
      "    CreateAtomBoolPropertyList(...)\n",
      "        CreateAtomBoolPropertyList( (Mol)mol, (str)propName [, (str)missingValueMarker='' [, (int)lineSize=190]]) -> None :\n",
      "            creates a list property on the molecule from individual atom property values\n",
      "        \n",
      "            C++ signature :\n",
      "                void CreateAtomBoolPropertyList(RDKit::ROMol {lvalue},std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='' [,unsigned int=190]])\n",
      "    \n",
      "    CreateAtomDoublePropertyList(...)\n",
      "        CreateAtomDoublePropertyList( (Mol)mol, (str)propName [, (str)missingValueMarker='' [, (int)lineSize=190]]) -> None :\n",
      "            creates a list property on the molecule from individual atom property values\n",
      "        \n",
      "            C++ signature :\n",
      "                void CreateAtomDoublePropertyList(RDKit::ROMol {lvalue},std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='' [,unsigned int=190]])\n",
      "    \n",
      "    CreateAtomIntPropertyList(...)\n",
      "        CreateAtomIntPropertyList( (Mol)mol, (str)propName [, (str)missingValueMarker='' [, (int)lineSize=190]]) -> None :\n",
      "            creates a list property on the molecule from individual atom property values\n",
      "        \n",
      "            C++ signature :\n",
      "                void CreateAtomIntPropertyList(RDKit::ROMol {lvalue},std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='' [,unsigned int=190]])\n",
      "    \n",
      "    CreateAtomStringPropertyList(...)\n",
      "        CreateAtomStringPropertyList( (Mol)mol, (str)propName [, (str)missingValueMarker='' [, (int)lineSize=190]]) -> None :\n",
      "            creates a list property on the molecule from individual atom property values\n",
      "        \n",
      "            C++ signature :\n",
      "                void CreateAtomStringPropertyList(RDKit::ROMol {lvalue},std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='' [,unsigned int=190]])\n",
      "    \n",
      "    CreateMolSubstanceGroup(...)\n",
      "        CreateMolSubstanceGroup( (Mol)mol, (str)type) -> SubstanceGroup :\n",
      "            creates a new SubstanceGroup associated with a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SubstanceGroup* CreateMolSubstanceGroup(RDKit::ROMol {lvalue},std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    CreateStereoGroup(...)\n",
      "        CreateStereoGroup( (StereoGroupType)stereoGroupType, (Mol)mol, (AtomPairsParameters)atomIds) -> StereoGroup :\n",
      "            creates a StereoGroup associated with a molecule from a list of atom Ids\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::StereoGroup* CreateStereoGroup(RDKit::StereoGroupType,RDKit::ROMol {lvalue},boost::python::api::object)\n",
      "    \n",
      "    DeleteSubstructs(...)\n",
      "        DeleteSubstructs( (Mol)mol, (Mol)query [, (bool)onlyFrags=False [, (bool)useChirality=False]]) -> Mol :\n",
      "            Removes atoms matching a substructure query from a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - query: the molecule to be used as a substructure query\n",
      "            \n",
      "                - onlyFrags: (optional) if this toggle is set, atoms will only be removed if\n",
      "                  the entire fragment in which they are found is matched by the query.\n",
      "                  See below for examples.\n",
      "                  Default value is 0 (remove the atoms whether or not the entire fragment matches)\n",
      "            \n",
      "                - useChirality: (optional) match the substructure query using chirality\n",
      "            \n",
      "              RETURNS: a new molecule with the substructure removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               The following examples substitute SMILES/SMARTS strings for molecules, you'd have\n",
      "               to actually use molecules:\n",
      "            \n",
      "                - DeleteSubstructs('CCOC','OC') -> 'CC'\n",
      "            \n",
      "                - DeleteSubstructs('CCOC','OC',1) -> 'CCOC'\n",
      "            \n",
      "                - DeleteSubstructs('CCOCCl.Cl','Cl',1) -> 'CCOCCl'\n",
      "            \n",
      "                - DeleteSubstructs('CCOCCl.Cl','Cl') -> 'CCOC'\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* DeleteSubstructs(RDKit::ROMol,RDKit::ROMol [,bool=False [,bool=False]])\n",
      "    \n",
      "    DetectBondStereoChemistry(...)\n",
      "        DetectBondStereoChemistry( (Mol)mol, (Conformer)conformer) -> None :\n",
      "            Assign stereochemistry to bonds based on coordinates and a conformer.\n",
      "                    DEPRECATED\n",
      "                    \n",
      "              ARGUMENTS:\n",
      "              \n",
      "                - mol: the molecule to be modified\n",
      "                - conformer: Conformer providing the coordinates\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void DetectBondStereoChemistry(RDKit::ROMol {lvalue},RDKit::Conformer const*)\n",
      "    \n",
      "    DetectBondStereochemistry(...)\n",
      "        DetectBondStereochemistry( (Mol)mol [, (int)confId=-1]) -> None :\n",
      "            DEPRECATED, use SetDoubleBondNeighborDirections() instead\n",
      "              ARGUMENTS:\n",
      "              \n",
      "                - mol: the molecule to be modified\n",
      "                - confId: Conformer to use for the coordinates\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void DetectBondStereochemistry(RDKit::ROMol {lvalue} [,int=-1])\n",
      "    \n",
      "    DetectChemistryProblems(...)\n",
      "        DetectChemistryProblems( (Mol)mol [, (int)sanitizeOps=rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL]) -> tuple :\n",
      "            checks for chemistry problems\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple DetectChemistryProblems(RDKit::ROMol [,unsigned int=rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL])\n",
      "    \n",
      "    FastFindRings(...)\n",
      "        FastFindRings( (Mol)arg1) -> None :\n",
      "            Does a non-SSSR ring finding for a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use.\n",
      "            \n",
      "              RETURNS: Nothing\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void FastFindRings(RDKit::ROMol)\n",
      "    \n",
      "    FindAllPathsOfLengthN(...)\n",
      "        FindAllPathsOfLengthN( (Mol)mol, (int)length [, (bool)useBonds=True [, (bool)useHs=False [, (int)rootedAtAtom=-1]]]) -> _listNSt3__16vectorIiNS_9allocatorIiEEEE :\n",
      "            Finds all paths of a particular length in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - length: an integer with the target length for the paths.\n",
      "            \n",
      "                - useBonds: (optional) toggles the use of bond indices in the paths.\n",
      "                  Otherwise atom indices are used.  *Note* this behavior is different\n",
      "                  from that for subgraphs.\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - rootedAtAtom: (optional) if nonzero, only paths from the specified\n",
      "                  atom will be returned.\n",
      "            \n",
      "              RETURNS: a tuple of tuples with IDs for the bonds.\n",
      "            \n",
      "              NOTES: \n",
      "            \n",
      "               - Difference between _subgraphs_ and _paths_ :: \n",
      "            \n",
      "                   Subgraphs are potentially branched, whereas paths (in our \n",
      "                   terminology at least) cannot be.  So, the following graph: \n",
      "            \n",
      "                        C--0--C--1--C--3--C\n",
      "                              |\n",
      "                              2\n",
      "                              |\n",
      "                              C\n",
      "            \n",
      "                   has 3 _subgraphs_ of length 3: (0,1,2),(0,1,3),(2,1,3)\n",
      "                   but only 2 _paths_ of length 3: (0,1,3),(2,1,3)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::list<std::__1::vector<int, std::__1::allocator<int> >, std::__1::allocator<std::__1::vector<int, std::__1::allocator<int> > > > FindAllPathsOfLengthN(RDKit::ROMol,unsigned int [,bool=True [,bool=False [,int=-1]]])\n",
      "    \n",
      "    FindAllSubgraphsOfLengthMToN(...)\n",
      "        FindAllSubgraphsOfLengthMToN( (Mol)mol, (int)min, (int)max [, (bool)useHs=False [, (int)rootedAtAtom=-1]]) -> object :\n",
      "            Finds all subgraphs of a particular length in a molecule\n",
      "              See documentation for FindAllSubgraphsOfLengthN for definitions\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::api::object FindAllSubgraphsOfLengthMToN(RDKit::ROMol,unsigned int,unsigned int [,bool=False [,int=-1]])\n",
      "    \n",
      "    FindAllSubgraphsOfLengthN(...)\n",
      "        FindAllSubgraphsOfLengthN( (Mol)mol, (int)length [, (bool)useHs=False [, (int)rootedAtAtom=-1]]) -> _listNSt3__16vectorIiNS_9allocatorIiEEEE :\n",
      "            Finds all subgraphs of a particular length in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - length: an integer with the target number of bonds for the subgraphs.\n",
      "            \n",
      "                - useHs: (optional) toggles whether or not bonds to Hs that are part of the graph\n",
      "                  should be included in the results.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - rootedAtAtom: (optional) if nonzero, only subgraphs from the specified\n",
      "                  atom will be returned.\n",
      "            \n",
      "              RETURNS: a tuple of 2-tuples with bond IDs\n",
      "            \n",
      "              NOTES: \n",
      "            \n",
      "               - Difference between _subgraphs_ and _paths_ :: \n",
      "            \n",
      "                   Subgraphs are potentially branched, whereas paths (in our \n",
      "                   terminology at least) cannot be.  So, the following graph: \n",
      "            \n",
      "                        C--0--C--1--C--3--C\n",
      "                              |\n",
      "                              2\n",
      "                              |\n",
      "                              C\n",
      "              has 3 _subgraphs_ of length 3: (0,1,2),(0,1,3),(2,1,3)\n",
      "              but only 2 _paths_ of length 3: (0,1,3),(2,1,3)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::list<std::__1::vector<int, std::__1::allocator<int> >, std::__1::allocator<std::__1::vector<int, std::__1::allocator<int> > > > FindAllSubgraphsOfLengthN(RDKit::ROMol,unsigned int [,bool=False [,int=-1]])\n",
      "    \n",
      "    FindAtomEnvironmentOfRadiusN(...)\n",
      "        FindAtomEnvironmentOfRadiusN( (Mol)mol, (int)radius, (int)rootedAtAtom [, (bool)useHs=False]) -> _vecti :\n",
      "            Finds the bonds within a certain radius of an atom in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - radius: an integer with the target radius for the environment.\n",
      "            \n",
      "                - rootedAtAtom: the atom to consider\n",
      "            \n",
      "                - useHs: (optional) toggles whether or not bonds to Hs that are part of the graph\n",
      "                  should be included in the results.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "              RETURNS: a vector of bond IDs\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::vector<int, std::__1::allocator<int> > FindAtomEnvironmentOfRadiusN(RDKit::ROMol,unsigned int,unsigned int [,bool=False])\n",
      "    \n",
      "    FindMolChiralCenters(mol, force=True, includeUnassigned=False)\n",
      "        >>> from rdkit import Chem\n",
      "        >>> mol = Chem.MolFromSmiles('[C@H](Cl)(F)Br')\n",
      "        >>> FindMolChiralCenters(mol)\n",
      "        [(0, 'R')]\n",
      "        >>> mol = Chem.MolFromSmiles('[C@@H](Cl)(F)Br')\n",
      "        >>> FindMolChiralCenters(mol)\n",
      "        [(0, 'S')]\n",
      "        \n",
      "        >>> FindMolChiralCenters(Chem.MolFromSmiles('CCC'))\n",
      "        []\n",
      "        \n",
      "        By default unassigned stereo centers are not reported:\n",
      "        \n",
      "        >>> mol = Chem.MolFromSmiles('C[C@H](F)C(F)(Cl)Br')\n",
      "        >>> FindMolChiralCenters(mol,force=True)\n",
      "        [(1, 'S')]\n",
      "        \n",
      "        but this can be changed:\n",
      "        \n",
      "        >>> FindMolChiralCenters(mol,force=True,includeUnassigned=True)\n",
      "        [(1, 'S'), (3, '?')]\n",
      "        \n",
      "        The handling of unassigned stereocenters for dependent stereochemistry is not correct:\n",
      "        \n",
      "        >>> Chem.FindMolChiralCenters(Chem.MolFromSmiles('C1CC(C)C(C)C(C)C1'),includeUnassigned=True)\n",
      "        [(2, '?'), (6, '?')]\n",
      "        >>> Chem.FindMolChiralCenters(Chem.MolFromSmiles('C1C[C@H](C)C(C)[C@H](C)C1'),includeUnassigned=True)\n",
      "        [(2, 'S'), (4, '?'), (6, 'R')]\n",
      "    \n",
      "    FindPotentialStereoBonds(...)\n",
      "        FindPotentialStereoBonds( (Mol)mol [, (bool)cleanIt=False]) -> None :\n",
      "            Find bonds than can be cis/trans in a molecule and mark them as 'any'.\n",
      "                     This function finds any double bonds that can potentially be part\n",
      "                     of a cis/trans system. No attempt is made here to mark them cis or trans\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - cleanIt: (optional) if this option is set to true, any previous marking of _CIPCode\n",
      "                           on the bond is cleared - otherwise it is left untouched\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void FindPotentialStereoBonds(RDKit::ROMol {lvalue} [,bool=False])\n",
      "    \n",
      "    FindRingFamilies(...)\n",
      "        FindRingFamilies( (Mol)arg1) -> None :\n",
      "            generate Unique Ring Families\n",
      "        \n",
      "            C++ signature :\n",
      "                void FindRingFamilies(RDKit::ROMol)\n",
      "    \n",
      "    FindUniqueSubgraphsOfLengthN(...)\n",
      "        FindUniqueSubgraphsOfLengthN( (Mol)mol, (int)length [, (bool)useHs=False [, (bool)useBO=True [, (int)rootedAtAtom=-1]]]) -> _listNSt3__16vectorIiNS_9allocatorIiEEEE :\n",
      "            Finds unique subgraphs of a particular length in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - length: an integer with the target number of bonds for the subgraphs.\n",
      "            \n",
      "                - useHs: (optional) toggles whether or not bonds to Hs that are part of the graph\n",
      "                  should be included in the results.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - useBO: (optional) Toggles use of bond orders in distinguishing one subgraph from\n",
      "                  another.\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - rootedAtAtom: (optional) if nonzero, only subgraphs from the specified\n",
      "                  atom will be returned.\n",
      "            \n",
      "              RETURNS: a tuple of tuples with bond IDs\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::list<std::__1::vector<int, std::__1::allocator<int> >, std::__1::allocator<std::__1::vector<int, std::__1::allocator<int> > > > FindUniqueSubgraphsOfLengthN(RDKit::ROMol,unsigned int [,bool=False [,bool=True [,int=-1]]])\n",
      "    \n",
      "    FragmentOnBRICSBonds(...)\n",
      "        FragmentOnBRICSBonds( (Mol)mol) -> Mol :\n",
      "            Return a new molecule with all BRICS bonds broken\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* FragmentOnBRICSBonds(RDKit::ROMol)\n",
      "    \n",
      "    FragmentOnBonds(...)\n",
      "        FragmentOnBonds( (Mol)mol, (AtomPairsParameters)bondIndices [, (bool)addDummies=True [, (AtomPairsParameters)dummyLabels=None [, (AtomPairsParameters)bondTypes=None [, (list)cutsPerAtom=[]]]]]) -> Mol :\n",
      "            Return a new molecule with all specified bonds broken\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                  - mol            - the molecule to be modified\n",
      "                  - bondIndices    - indices of the bonds to be broken\n",
      "                  - addDummies  - toggles addition of dummy atoms to indicate where \n",
      "                    bonds were broken\n",
      "                  - dummyLabels - used to provide the labels to be used for the dummies.\n",
      "                    the first element in each pair is the label for the dummy\n",
      "                    that replaces the bond's beginAtom, the second is for the \n",
      "                    dummy that replaces the bond's endAtom. If not provided, the\n",
      "                    dummies are labeled with atom indices.\n",
      "                  - bondTypes - used to provide the bond type to use between the\n",
      "                    fragments and the dummy atoms. If not provided, defaults to single. \n",
      "                  - cutsPerAtom - used to return the number of cuts made at each atom. \n",
      "            \n",
      "              RETURNS:\n",
      "                  a new Mol with the modifications\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* FragmentOnBonds(RDKit::ROMol,boost::python::api::object [,bool=True [,boost::python::api::object=None [,boost::python::api::object=None [,boost::python::list=[]]]]])\n",
      "    \n",
      "    FragmentOnSomeBonds(...)\n",
      "        FragmentOnSomeBonds( (Mol)mol, (AtomPairsParameters)bondIndices [, (int)numToBreak=1 [, (bool)addDummies=True [, (AtomPairsParameters)dummyLabels=None [, (AtomPairsParameters)bondTypes=None [, (bool)returnCutsPerAtom=False]]]]]) -> tuple :\n",
      "            fragment on some bonds\n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple FragmentOnSomeBonds(RDKit::ROMol,boost::python::api::object [,unsigned int=1 [,bool=True [,boost::python::api::object=None [,boost::python::api::object=None [,bool=False]]]]])\n",
      "    \n",
      "    Get3DDistanceMatrix(...)\n",
      "        Get3DDistanceMatrix( (Mol)mol [, (int)confId=-1 [, (bool)useAtomWts=False [, (bool)force=False [, (str)prefix='']]]]) -> object :\n",
      "            Returns the molecule's 3D distance matrix.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - confId: (optional) chooses the conformer Id to use\n",
      "                  Default value is -1.\n",
      "            \n",
      "                - useAtomWts: (optional) toggles using atom weights for the diagonal elements of the\n",
      "                  matrix (to return a \"Balaban\" distance matrix).\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - force: (optional) forces the calculation to proceed, even if there is a cached value.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - prefix: (optional, internal use) sets the prefix used in the property cache\n",
      "                  Default value is .\n",
      "            \n",
      "              RETURNS: a Numeric array of floats with the distance matrix\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* Get3DDistanceMatrix(RDKit::ROMol {lvalue} [,int=-1 [,bool=False [,bool=False [,char const*='']]]])\n",
      "    \n",
      "    GetAdjacencyMatrix(...)\n",
      "        GetAdjacencyMatrix( (Mol)mol [, (bool)useBO=False [, (int)emptyVal=0 [, (bool)force=False [, (str)prefix='']]]]) -> object :\n",
      "            Returns the molecule's adjacency matrix.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - useBO: (optional) toggles use of bond orders in calculating the matrix.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - emptyVal: (optional) sets the elements of the matrix between non-adjacent atoms\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - force: (optional) forces the calculation to proceed, even if there is a cached value.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - prefix: (optional, internal use) sets the prefix used in the property cache\n",
      "                  Default value is .\n",
      "            \n",
      "              RETURNS: a Numeric array of floats containing the adjacency matrix\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetAdjacencyMatrix(RDKit::ROMol {lvalue} [,bool=False [,int=0 [,bool=False [,char const*='']]]])\n",
      "    \n",
      "    GetAtomAlias(...)\n",
      "        GetAtomAlias( (Atom)atom) -> str :\n",
      "            Returns the atom's MDL alias text\n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > GetAtomAlias(RDKit::Atom const*)\n",
      "    \n",
      "    GetAtomRLabel(...)\n",
      "        GetAtomRLabel( (Atom)atom) -> int :\n",
      "            Returns the atom's MDL AtomRLabel (this is an integer from 0 to 99)\n",
      "        \n",
      "            C++ signature :\n",
      "                int GetAtomRLabel(RDKit::Atom const*)\n",
      "    \n",
      "    GetAtomValue(...)\n",
      "        GetAtomValue( (Atom)atom) -> str :\n",
      "            Returns the atom's MDL alias text\n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > GetAtomValue(RDKit::Atom const*)\n",
      "    \n",
      "    GetDefaultPickleProperties(...)\n",
      "        GetDefaultPickleProperties() -> int :\n",
      "            Get the current global mol pickler options.\n",
      "        \n",
      "            C++ signature :\n",
      "                unsigned int GetDefaultPickleProperties()\n",
      "    \n",
      "    GetDistanceMatrix(...)\n",
      "        GetDistanceMatrix( (Mol)mol [, (bool)useBO=False [, (bool)useAtomWts=False [, (bool)force=False [, (str)prefix='']]]]) -> object :\n",
      "            Returns the molecule's topological distance matrix.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - useBO: (optional) toggles use of bond orders in calculating the distance matrix.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - useAtomWts: (optional) toggles using atom weights for the diagonal elements of the\n",
      "                  matrix (to return a \"Balaban\" distance matrix).\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - force: (optional) forces the calculation to proceed, even if there is a cached value.\n",
      "                  Default value is 0.\n",
      "            \n",
      "                - prefix: (optional, internal use) sets the prefix used in the property cache\n",
      "                  Default value is .\n",
      "            \n",
      "              RETURNS: a Numeric array of floats with the distance matrix\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* GetDistanceMatrix(RDKit::ROMol {lvalue} [,bool=False [,bool=False [,bool=False [,char const*='']]]])\n",
      "    \n",
      "    GetFormalCharge(...)\n",
      "        GetFormalCharge( (Mol)arg1) -> int :\n",
      "            Returns the formal charge for the molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int GetFormalCharge(RDKit::ROMol)\n",
      "    \n",
      "    GetMolFrags(...)\n",
      "        GetMolFrags( (Mol)mol [, (bool)asMols=False [, (bool)sanitizeFrags=True [, (AtomPairsParameters)frags=None [, (AtomPairsParameters)fragsMolAtomMapping=None]]]]) -> tuple :\n",
      "            Finds the disconnected fragments from a molecule.\n",
      "            \n",
      "              For example, for the molecule 'CC(=O)[O-].[NH3+]C' GetMolFrags() returns\n",
      "              ((0, 1, 2, 3), (4, 5))\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - asMols: (optional) if this is provided and true, the fragments\n",
      "                  will be returned as molecules instead of atom ids.\n",
      "                - sanitizeFrags: (optional) if this is provided and true, the fragments\n",
      "                  molecules will be sanitized before returning them.\n",
      "                - frags: (optional, defaults to None) if this is provided as an empty list,\n",
      "                  the result will be mol.GetNumAtoms() long on return and will contain the\n",
      "                  fragment assignment for each Atom\n",
      "                - fragsMolAtomMapping: (optional, defaults to None) if this is provided as\n",
      "                  an empty list, the result will be a a numFrags long list on return, and\n",
      "                  each entry will contain the indices of the Atoms in that fragment:\n",
      "                  [(0, 1, 2, 3), (4, 5)]\n",
      "            \n",
      "              RETURNS: a tuple of tuples with IDs for the atoms in each fragment\n",
      "                       or a tuple of molecules.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple GetMolFrags(RDKit::ROMol [,bool=False [,bool=True [,boost::python::api::object=None [,boost::python::api::object=None]]]])\n",
      "    \n",
      "    GetMolSubstanceGroupWithIdx(...)\n",
      "        GetMolSubstanceGroupWithIdx( (Mol)arg1, (int)arg2) -> SubstanceGroup :\n",
      "            returns a particular SubstanceGroup from the molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SubstanceGroup* GetMolSubstanceGroupWithIdx(RDKit::ROMol {lvalue},unsigned int)\n",
      "    \n",
      "    GetMolSubstanceGroups(...)\n",
      "        GetMolSubstanceGroups( (Mol)arg1) -> SubstanceGroup_VECT :\n",
      "            returns a copy of the molecule's SubstanceGroups (if any)\n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::vector<RDKit::SubstanceGroup, std::__1::allocator<RDKit::SubstanceGroup> > GetMolSubstanceGroups(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    GetPeriodicTable(...)\n",
      "        GetPeriodicTable() -> PeriodicTable :\n",
      "            Returns the application's PeriodicTable instance.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::PeriodicTable* GetPeriodicTable()\n",
      "    \n",
      "    GetSSSR(...)\n",
      "        GetSSSR( (Mol)arg1) -> int :\n",
      "            Get the smallest set of simple rings for a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use.\n",
      "            \n",
      "              RETURNS: a sequence of sequences containing the rings found as atom ids\n",
      "                     The length of this will be equal to NumBonds-NumAtoms+1 for single-fragment molecules.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                int GetSSSR(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    GetShortestPath(...)\n",
      "        GetShortestPath( (Mol)arg1, (int)arg2, (int)arg3) -> tuple :\n",
      "            Find the shortest path between two atoms using the Bellman-Ford algorithm.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - idx1: index of the first atom\n",
      "                - idx2: index of the second atom\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple GetShortestPath(RDKit::ROMol,int,int)\n",
      "    \n",
      "    GetSupplementalSmilesLabel(...)\n",
      "        GetSupplementalSmilesLabel( (Atom)atom) -> str :\n",
      "            Gets the supplemental smiles label on an atom, returns an empty string if not present.\n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > GetSupplementalSmilesLabel(RDKit::Atom const*)\n",
      "    \n",
      "    GetSymmSSSR(...)\n",
      "        GetSymmSSSR( (Mol)arg1) -> _vectNSt3__16vectorIiNS_9allocatorIiEEEE :\n",
      "            Get a symmetrized SSSR for a molecule.\n",
      "            \n",
      "              The symmetrized SSSR is at least as large as the SSSR for a molecule.\n",
      "              In certain highly-symmetric cases (e.g. cubane), the symmetrized SSSR can be\n",
      "              a bit larger (i.e. the number of symmetrized rings is >= NumBonds-NumAtoms+1).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use.\n",
      "            \n",
      "              RETURNS: a sequence of sequences containing the rings found as atom ids\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::vector<std::__1::vector<int, std::__1::allocator<int> >, std::__1::allocator<std::__1::vector<int, std::__1::allocator<int> > > > GetSymmSSSR(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    JSONToMols(...)\n",
      "        JSONToMols( (str)jsonBlock [, (AtomPairsParameters)params=None]) -> tuple :\n",
      "            Convert JSON to a tuple of molecules\n",
      "            \n",
      "                ARGUMENTS:\n",
      "                  - jsonBlock: the molecule to work with\n",
      "                  - params: (optional) JSONParseParameters controlling the JSON parsing\n",
      "                RETURNS:\n",
      "                  a tuple of Mols\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::tuple JSONToMols(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,boost::python::api::object=None])\n",
      "    \n",
      "    Kekulize(...)\n",
      "        Kekulize( (Mol)mol [, (bool)clearAromaticFlags=False]) -> None :\n",
      "            Kekulizes the molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - clearAromaticFlags: (optional) if this toggle is set, all atoms and bonds in the \n",
      "                  molecule will be marked non-aromatic following the kekulization.\n",
      "                  Default value is 0.\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void Kekulize(RDKit::ROMol {lvalue} [,bool=False])\n",
      "    \n",
      "    LayeredFingerprint(...)\n",
      "        LayeredFingerprint( (Mol)mol [, (int)layerFlags=4294967295 [, (int)minPath=1 [, (int)maxPath=7 [, (int)fpSize=2048 [, (list)atomCounts=[] [, (ExplicitBitVect)setOnlyBits=None [, (bool)branchedPaths=True [, (AtomPairsParameters)fromAtoms=0]]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns a layered fingerprint for a molecule\n",
      "            \n",
      "              NOTE: This function is experimental. The API or results may change from\n",
      "                release to release.\n",
      "            \n",
      "              Explanation of the algorithm below.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - layerFlags: (optional) which layers to include in the fingerprint\n",
      "                  See below for definitions. Defaults to all.\n",
      "            \n",
      "                - minPath: (optional) minimum number of bonds to include in the subgraphs\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - maxPath: (optional) maximum number of bonds to include in the subgraphs\n",
      "                  Defaults to 7.\n",
      "            \n",
      "                - fpSize: (optional) number of bits in the fingerprint\n",
      "                  Defaults to 2048.\n",
      "            \n",
      "                - atomCounts: (optional) \n",
      "                  if provided, this should be a list at least as long as the number of atoms\n",
      "                  in the molecule. It will be used to provide the count of the number \n",
      "                  of paths that set bits each atom is involved in.\n",
      "                  NOTE: the list is not zeroed out here.\n",
      "            \n",
      "                - setOnlyBits: (optional) \n",
      "                  if provided, only bits that are set in this bit vector will be set\n",
      "                  in the result. This is essentially the same as doing:\n",
      "                       res &= setOnlyBits\n",
      "                  but also has an impact on the atomCounts (if being used)\n",
      "            \n",
      "                - branchedPaths: (optional) if set both branched and unbranched paths will be\n",
      "                  used in the fingerprint.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - fromAtoms: (optional) a sequence of atom indices. If provided, only paths/subgraphs \n",
      "                  starting from these atoms will be used.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "              RETURNS: a DataStructs.ExplicitBitVect with _fpSize_ bits\n",
      "            \n",
      "              Layer definitions:\n",
      "                 - 0x01: pure topology\n",
      "                 - 0x02: bond order\n",
      "                 - 0x04: atom types\n",
      "                 - 0x08: presence of rings\n",
      "                 - 0x10: ring sizes\n",
      "                 - 0x20: aromaticity\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* LayeredFingerprint(RDKit::ROMol [,unsigned int=4294967295 [,unsigned int=1 [,unsigned int=7 [,unsigned int=2048 [,boost::python::list=[] [,ExplicitBitVect*=None [,bool=True [,boost::python::api::object=0]]]]]]]])\n",
      "    \n",
      "    LogErrorMsg(...)\n",
      "        LogErrorMsg( (str)arg1) -> None :\n",
      "            Log a warning message to the RDKit error logs\n",
      "        \n",
      "            C++ signature :\n",
      "                void LogErrorMsg(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    LogWarningMsg(...)\n",
      "        LogWarningMsg( (str)arg1) -> None :\n",
      "            Log a warning message to the RDKit warning logs\n",
      "        \n",
      "            C++ signature :\n",
      "                void LogWarningMsg(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    MergeQueryHs(...)\n",
      "        MergeQueryHs( (Mol)mol [, (bool)mergeUnmappedOnly=False]) -> Mol :\n",
      "            merges hydrogens into their neighboring atoms as queries\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MergeQueryHs(RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    MolAddRecursiveQueries(...)\n",
      "        MolAddRecursiveQueries( (Mol)mol, (dict)queries, (str)propName) -> None :\n",
      "            Adds named recursive queries to atoms\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolAddRecursiveQueries(RDKit::ROMol {lvalue},boost::python::dict,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    MolFragmentToCXSmiles(...)\n",
      "        MolFragmentToCXSmiles( (Mol)mol, (AtomPairsParameters)atomsToUse [, (AtomPairsParameters)bondsToUse=0 [, (AtomPairsParameters)atomSymbols=0 [, (AtomPairsParameters)bondSymbols=0 [, (bool)isomericSmiles=True [, (bool)kekuleSmiles=False [, (int)rootedAtAtom=-1 [, (bool)canonical=True [, (bool)allBondsExplicit=False [, (bool)allHsExplicit=False]]]]]]]]]) -> str :\n",
      "            Returns the CXSMILES string for a fragment of a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - atomsToUse : a list of atoms to include in the fragment\n",
      "                - bondsToUse : (optional) a list of bonds to include in the fragment\n",
      "                  if not provided, all bonds between the atoms provided\n",
      "                  will be included.\n",
      "                - atomSymbols : (optional) a list with the symbols to use for the atoms\n",
      "                  in the SMILES. This should have be mol.GetNumAtoms() long.\n",
      "                - bondSymbols : (optional) a list with the symbols to use for the bonds\n",
      "                  in the SMILES. This should have be mol.GetNumBonds() long.\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMILES.  Defaults to true.\n",
      "                - kekuleSmiles: (optional) use the Kekule form (no aromatic bonds) in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - rootedAtAtom: (optional) if non-negative, this forces the SMILES \n",
      "                  to start at a particular atom. Defaults to -1.\n",
      "                - canonical: (optional) if false no attempt will be made to canonicalize\n",
      "                  the molecule. Defaults to true.\n",
      "                - allBondsExplicit: (optional) if true, all bond orders will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - allHsExplicit: (optional) if true, all H counts will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - doRandom: (optional) if true, randomized the DFS transversal graph,\n",
      "                  so we can generate random smiles. Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolFragmentToCXSmiles(RDKit::ROMol,boost::python::api::object [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=True [,bool=False [,int=-1 [,bool=True [,bool=False [,bool=False]]]]]]]]])\n",
      "    \n",
      "    MolFragmentToSmarts(...)\n",
      "        MolFragmentToSmarts( (Mol)mol, (AtomPairsParameters)atomsToUse [, (AtomPairsParameters)bondsToUse=0 [, (bool)isomericSmarts=True]]) -> str :\n",
      "            Returns a SMARTS string for a fragment of a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - atomsToUse: indices of atoms to include in the SMARTS string\n",
      "                - bondsToUse: indices of bonds to include in the SMARTS string (optional)\n",
      "                - isomericSmarts: (optional) include information about stereochemistry in\n",
      "                  the SMARTS.  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolFragmentToSmarts(RDKit::ROMol,boost::python::api::object [,boost::python::api::object=0 [,bool=True]])\n",
      "    \n",
      "    MolFragmentToSmiles(...)\n",
      "        MolFragmentToSmiles( (Mol)mol, (AtomPairsParameters)atomsToUse [, (AtomPairsParameters)bondsToUse=0 [, (AtomPairsParameters)atomSymbols=0 [, (AtomPairsParameters)bondSymbols=0 [, (bool)isomericSmiles=True [, (bool)kekuleSmiles=False [, (int)rootedAtAtom=-1 [, (bool)canonical=True [, (bool)allBondsExplicit=False [, (bool)allHsExplicit=False]]]]]]]]]) -> str :\n",
      "            Returns the canonical SMILES string for a fragment of a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - atomsToUse : a list of atoms to include in the fragment\n",
      "                - bondsToUse : (optional) a list of bonds to include in the fragment\n",
      "                  if not provided, all bonds between the atoms provided\n",
      "                  will be included.\n",
      "                - atomSymbols : (optional) a list with the symbols to use for the atoms\n",
      "                  in the SMILES. This should have be mol.GetNumAtoms() long.\n",
      "                - bondSymbols : (optional) a list with the symbols to use for the bonds\n",
      "                  in the SMILES. This should have be mol.GetNumBonds() long.\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMILES.  Defaults to true.\n",
      "                - kekuleSmiles: (optional) use the Kekule form (no aromatic bonds) in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - rootedAtAtom: (optional) if non-negative, this forces the SMILES \n",
      "                  to start at a particular atom. Defaults to -1.\n",
      "                - canonical: (optional) if false no attempt will be made to canonicalize\n",
      "                  the molecule. Defaults to true.\n",
      "                - allBondsExplicit: (optional) if true, all bond orders will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - allHsExplicit: (optional) if true, all H counts will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - doRandom: (optional) if true, randomized the DFS transversal graph,\n",
      "                  so we can generate random smiles. Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolFragmentToSmiles(RDKit::ROMol,boost::python::api::object [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=0 [,bool=True [,bool=False [,int=-1 [,bool=True [,bool=False [,bool=False]]]]]]]]])\n",
      "    \n",
      "    MolFromFASTA(...)\n",
      "        MolFromFASTA( (AtomPairsParameters)text [, (bool)sanitize=True [, (int)flavor=0]]) -> Mol :\n",
      "            Construct a molecule from a FASTA string (currently only supports peptides).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - text: string containing the FASTA\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "            - flavor: (optional)\n",
      "                - 0 Protein, L amino acids (default)\n",
      "                - 1 Protein, D amino acids\n",
      "                - 2 RNA, no cap\n",
      "                - 3 RNA, 5' cap\n",
      "                - 4 RNA, 3' cap\n",
      "                - 5 RNA, both caps\n",
      "                - 6 DNA, no cap\n",
      "                - 7 DNA, 5' cap\n",
      "                - 8 DNA, 3' cap\n",
      "                - 9 DNA, both caps\n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromFASTA(boost::python::api::object [,bool=True [,int=0]])\n",
      "    \n",
      "    MolFromHELM(...)\n",
      "        MolFromHELM( (AtomPairsParameters)text [, (bool)sanitize=True]) -> Mol :\n",
      "            Construct a molecule from a HELM string (currently only supports peptides).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - text: string containing the HELM\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromHELM(boost::python::api::object [,bool=True])\n",
      "    \n",
      "    MolFromMol2Block(...)\n",
      "        MolFromMol2Block( (str)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)cleanupSubstructures=True]]]) -> Mol :\n",
      "            Construct a molecule from a Tripos Mol2 block.\n",
      "            \n",
      "              NOTE:\n",
      "                The parser expects the atom-typing scheme used by Corina.\n",
      "                Atom types from Tripos' dbtranslate are less supported.\n",
      "                Other atom typing schemes are unlikely to work.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol2Block: string containing the Mol2 block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - cleanupSubstructures: (optional) toggles standardizing some \n",
      "                  substructures found in mol2 files.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMol2Block(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    MolFromMol2File(...)\n",
      "        MolFromMol2File( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)cleanupSubstructures=True]]]) -> Mol :\n",
      "            Construct a molecule from a Tripos Mol2 file.\n",
      "            \n",
      "              NOTE:\n",
      "                The parser expects the atom-typing scheme used by Corina.\n",
      "                Atom types from Tripos' dbtranslate are less supported.\n",
      "                Other atom typing schemes are unlikely to work.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "                                              \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - cleanupSubstructures: (optional) toggles standardizing some \n",
      "                  substructures found in mol2 files.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMol2File(char const* [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    MolFromMolBlock(...)\n",
      "        MolFromMolBlock( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molBlock: string containing the Mol block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolBlock(boost::python::api::object [,bool=True [,bool=True [,bool=True]]])\n",
      "        \n",
      "        MolFromMolBlock( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molBlock: string containing the Mol block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolBlock(boost::python::api::object [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    MolFromMolFile(...)\n",
      "        MolFromMolFile( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolFile(char const* [,bool=True [,bool=True [,bool=True]]])\n",
      "        \n",
      "        MolFromMolFile( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (bool)strictParsing=True]]]) -> Mol :\n",
      "            Construct a molecule from a Mol file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - strictParsing: (optional) if this is false, the parser is more lax about.\n",
      "                  correctness of the content.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromMolFile(char const* [,bool=True [,bool=True [,bool=True]]])\n",
      "    \n",
      "    MolFromPDBBlock(...)\n",
      "        MolFromPDBBlock( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True [, (int)flavor=0 [, (bool)proximityBonding=True]]]]) -> Mol :\n",
      "            Construct a molecule from a PDB block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molBlock: string containing the PDB block\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - flavor: (optional) \n",
      "            \n",
      "                - proximityBonding: (optional) toggles automatic proximity bonding\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromPDBBlock(boost::python::api::object [,bool=True [,bool=True [,unsigned int=0 [,bool=True]]]])\n",
      "    \n",
      "    MolFromPDBFile(...)\n",
      "        MolFromPDBFile( (str)molFileName [, (bool)sanitize=True [, (bool)removeHs=True [, (int)flavor=0 [, (bool)proximityBonding=True]]]]) -> Mol :\n",
      "            Construct a molecule from a PDB file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "                - flavor: (optional) \n",
      "            \n",
      "                - proximityBonding: (optional) toggles automatic proximity bonding\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromPDBFile(char const* [,bool=True [,bool=True [,unsigned int=0 [,bool=True]]]])\n",
      "    \n",
      "    MolFromRDKitSVG(...)\n",
      "        MolFromRDKitSVG( (AtomPairsParameters)molBlock [, (bool)sanitize=True [, (bool)removeHs=True]]) -> Mol :\n",
      "            Construct a molecule from an RDKit-generate SVG string.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - svg: string containing the SVG data (must include molecule metadata)\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - removeHs: (optional) toggles removing hydrogens from the molecule.\n",
      "                  This only make sense when sanitization is done.\n",
      "                  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "              NOTE: this functionality should be considered beta.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromRDKitSVG(boost::python::api::object [,bool=True [,bool=True]])\n",
      "    \n",
      "    MolFromSequence(...)\n",
      "        MolFromSequence( (AtomPairsParameters)text [, (bool)sanitize=True [, (int)flavor=0]]) -> Mol :\n",
      "            Construct a molecule from a sequence string (currently only supports peptides).\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - text: string containing the sequence\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - flavor: (optional)\n",
      "                    - 0 Protein, L amino acids (default)\n",
      "                    - 1 Protein, D amino acids\n",
      "                    - 2 RNA, no cap\n",
      "                    - 3 RNA, 5' cap\n",
      "                    - 4 RNA, 3' cap\n",
      "                    - 5 RNA, both caps\n",
      "                    - 6 DNA, no cap\n",
      "                    - 7 DNA, 5' cap\n",
      "                    - 8 DNA, 3' cap\n",
      "                    - 9 DNA, both caps\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSequence(boost::python::api::object [,bool=True [,int=0]])\n",
      "    \n",
      "    MolFromSmarts(...)\n",
      "        MolFromSmarts( (AtomPairsParameters)SMARTS [, (bool)mergeHs=False [, (dict)replacements={}]]) -> Mol :\n",
      "            Construct a molecule from a SMARTS string.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - SMARTS: the smarts string\n",
      "            \n",
      "                - mergeHs: (optional) toggles the merging of explicit Hs in the query into the attached\n",
      "                  atoms.  So, for example, 'C[H]' becomes '[C;!H0]'.\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - replacements: (optional) a dictionary of replacement strings (see below)\n",
      "                  Defaults to {}. See the documentation for MolFromSmiles for an explanation.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSmarts(boost::python::api::object [,bool=False [,boost::python::dict={}]])\n",
      "    \n",
      "    MolFromSmiles(...)\n",
      "        MolFromSmiles( (AtomPairsParameters)SMILES, (SmilesParserParams)params) -> Mol :\n",
      "            Construct a molecule from a SMILES string.\n",
      "            \n",
      "                 ARGUMENTS:\n",
      "               \n",
      "                   - SMILES: the smiles string\n",
      "               \n",
      "                   - params: used to provide optional parameters for the SMILES parsing\n",
      "               \n",
      "                 RETURNS:\n",
      "               \n",
      "                   a Mol object, None on failure.\n",
      "               \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSmiles(boost::python::api::object,RDKit::SmilesParserParams)\n",
      "        \n",
      "        MolFromSmiles( (AtomPairsParameters)SMILES [, (bool)sanitize=True [, (dict)replacements={}]]) -> Mol :\n",
      "            Construct a molecule from a SMILES string.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - SMILES: the smiles string\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - replacements: (optional) a dictionary of replacement strings (see below)\n",
      "                  Defaults to {}.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "               The optional replacements dict can be used to do string substitution of abbreviations \n",
      "               in the input SMILES. The set of substitutions is repeatedly looped through until \n",
      "               the string no longer changes. It is the responsiblity of the caller to make sure \n",
      "               that substitutions results in legal and sensible SMILES. \n",
      "             \n",
      "               Examples of replacements: \n",
      "             \n",
      "                 CC{Q}C with {'{Q}':'OCCO'} -> CCOCCOC  \n",
      "            \n",
      "                 C{A}C{Q}C with {'{Q}':'OCCO', '{A}':'C1(CC1)'} -> CC1(CC1)COCCOC  \n",
      "            \n",
      "                 C{A}C{Q}C with {'{Q}':'{X}CC{X}', '{A}':'C1CC1', '{X}':'N'} -> CC1CC1CCNCCNC  \n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromSmiles(boost::python::api::object [,bool=True [,boost::python::dict={}]])\n",
      "    \n",
      "    MolFromTPLBlock(...)\n",
      "        MolFromTPLBlock( (AtomPairsParameters)tplBlock [, (bool)sanitize=True [, (bool)skipFirstConf=False]]) -> Mol :\n",
      "            Construct a molecule from a TPL block.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - skipFirstConf: (optional) skips reading the first conformer.\n",
      "                  Defaults to False.\n",
      "                  This should be set to True when reading TPLs written by \n",
      "                  the CombiCode.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromTPLBlock(boost::python::api::object [,bool=True [,bool=False]])\n",
      "    \n",
      "    MolFromTPLFile(...)\n",
      "        MolFromTPLFile( (str)fileName [, (bool)sanitize=True [, (bool)skipFirstConf=False]]) -> Mol :\n",
      "            Construct a molecule from a TPL file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - fileName: name of the file to read\n",
      "            \n",
      "                - sanitize: (optional) toggles sanitization of the molecule.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - skipFirstConf: (optional) skips reading the first conformer.\n",
      "                  Defaults to False.\n",
      "                  This should be set to True when reading TPLs written by \n",
      "                  the CombiCode.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a Mol object, None on failure.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MolFromTPLFile(char const* [,bool=True [,bool=False]])\n",
      "    \n",
      "    MolToCXSmiles(...)\n",
      "        MolToCXSmiles( (Mol)mol [, (bool)isomericSmiles=True [, (bool)kekuleSmiles=False [, (int)rootedAtAtom=-1 [, (bool)canonical=True [, (bool)allBondsExplicit=False [, (bool)allHsExplicit=False [, (bool)doRandom=False]]]]]]]) -> str :\n",
      "            Returns the CXSMILES string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMILES.  Defaults to true.\n",
      "                - kekuleSmiles: (optional) use the Kekule form (no aromatic bonds) in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - rootedAtAtom: (optional) if non-negative, this forces the SMILES \n",
      "                  to start at a particular atom. Defaults to -1.\n",
      "                - canonical: (optional) if false no attempt will be made to canonicalize\n",
      "                  the molecule. Defaults to true.\n",
      "                - allBondsExplicit: (optional) if true, all bond orders will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - allHsExplicit: (optional) if true, all H counts will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToCXSmiles(RDKit::ROMol [,bool=True [,bool=False [,int=-1 [,bool=True [,bool=False [,bool=False [,bool=False]]]]]]])\n",
      "    \n",
      "    MolToFASTA(...)\n",
      "        MolToFASTA( (Mol)mol) -> str :\n",
      "            Returns the FASTA string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "            \n",
      "              NOTE: the molecule should contain monomer information in AtomMonomerInfo structures \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToFASTA(RDKit::ROMol)\n",
      "    \n",
      "    MolToHELM(...)\n",
      "        MolToHELM( (Mol)mol) -> str :\n",
      "            Returns the HELM string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "            \n",
      "              NOTE: the molecule should contain monomer information in AtomMonomerInfo structures \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToHELM(RDKit::ROMol)\n",
      "    \n",
      "    MolToJSON(...)\n",
      "        MolToJSON( (Mol)mol) -> str :\n",
      "            Convert a single molecule to JSON\n",
      "            \n",
      "                ARGUMENTS:\n",
      "                  - mol: the molecule to work with\n",
      "                RETURNS:\n",
      "                  a string\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToJSON(RDKit::ROMol)\n",
      "    \n",
      "    MolToMolBlock(...)\n",
      "        MolToMolBlock( (Mol)mol [, (bool)includeStereo=True [, (int)confId=-1 [, (bool)kekulize=True [, (bool)forceV3000=False]]]]) -> str :\n",
      "            Returns a Mol block for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - includeStereo: (optional) toggles inclusion of stereochemical\n",
      "                  information in the output\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - kekulize: (optional) triggers kekulization of the molecule before it's written,\n",
      "                  as suggested by the MDL spec.\n",
      "                - forceV3000 (optional) force generation a V3000 mol block (happens automatically with \n",
      "                  more than 999 atoms or bonds)\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToMolBlock(RDKit::ROMol [,bool=True [,int=-1 [,bool=True [,bool=False]]]])\n",
      "    \n",
      "    MolToMolFile(...)\n",
      "        MolToMolFile( (Mol)mol, (str)filename [, (bool)includeStereo=True [, (int)confId=-1 [, (bool)kekulize=True [, (bool)forceV3000=False]]]]) -> None :\n",
      "            Writes a Mol file for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - filename: the file to write to\n",
      "                - includeStereo: (optional) toggles inclusion of stereochemical\n",
      "                  information in the output\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - kekulize: (optional) triggers kekulization of the molecule before it's written,\n",
      "                  as suggested by the MDL spec.\n",
      "                - forceV3000 (optional) force generation a V3000 mol block (happens automatically with \n",
      "                  more than 999 atoms or bonds)\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToMolFile(RDKit::ROMol,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,bool=True [,int=-1 [,bool=True [,bool=False]]]])\n",
      "    \n",
      "    MolToPDBBlock(...)\n",
      "        MolToPDBBlock( (Mol)mol [, (int)confId=-1 [, (int)flavor=0]]) -> str :\n",
      "            Returns a PDB block for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - flavor: (optional) \n",
      "                        - flavor & 1 : Write MODEL/ENDMDL lines around each record \n",
      "                        - flavor & 2 : Don't write any CONECT records \n",
      "                        - flavor & 4 : Write CONECT records in both directions \n",
      "                        - flavor & 8 : Don't use multiple CONECTs to encode bond order \n",
      "                        - flavor & 16 : Write MASTER record \n",
      "                        - flavor & 32 : Write TER record \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToPDBBlock(RDKit::ROMol [,int=-1 [,unsigned int=0]])\n",
      "    \n",
      "    MolToPDBFile(...)\n",
      "        MolToPDBFile( (Mol)mol, (str)filename [, (int)confId=-1 [, (int)flavor=0]]) -> None :\n",
      "            Writes a PDB file for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - filename: name of the file to write\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "                - flavor: (optional) \n",
      "                        - flavor & 1 : Write MODEL/ENDMDL lines around each record \n",
      "                        - flavor & 2 : Don't write any CONECT records \n",
      "                        - flavor & 4 : Write CONECT records in both directions \n",
      "                        - flavor & 8 : Don't use multiple CONECTs to encode bond order \n",
      "                        - flavor & 16 : Write MASTER record \n",
      "                        - flavor & 32 : Write TER record \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToPDBFile(RDKit::ROMol,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,int=-1 [,unsigned int=0]])\n",
      "    \n",
      "    MolToSVG(...)\n",
      "        MolToSVG( (Mol)arg1, (int)mol [, (int)width=300 [, (AtomPairsParameters)height=300 [, (bool)highlightAtoms=None [, (int)kekulize=True [, (int)lineWidthMult=1 [, (bool)fontSize=12 [, (int)includeAtomCircles=True]]]]]]]) -> str :\n",
      "            Returns svg for a molecule\n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToSVG(RDKit::ROMol,unsigned int [,unsigned int=300 [,boost::python::api::object=300 [,bool=None [,unsigned int=True [,unsigned int=1 [,bool=12 [,int=True]]]]]]])\n",
      "    \n",
      "    MolToSequence(...)\n",
      "        MolToSequence( (Mol)mol) -> str :\n",
      "            Returns the sequence string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "            \n",
      "              NOTE: the molecule should contain monomer information in AtomMonomerInfo structures \n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToSequence(RDKit::ROMol)\n",
      "    \n",
      "    MolToSmarts(...)\n",
      "        MolToSmarts( (Mol)mol [, (bool)isomericSmiles=True]) -> str :\n",
      "            Returns a SMARTS string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMARTS.  Defaults to true.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToSmarts(RDKit::ROMol [,bool=True])\n",
      "    \n",
      "    MolToSmiles(...)\n",
      "        MolToSmiles( (Mol)mol [, (bool)isomericSmiles=True [, (bool)kekuleSmiles=False [, (int)rootedAtAtom=-1 [, (bool)canonical=True [, (bool)allBondsExplicit=False [, (bool)allHsExplicit=False [, (bool)doRandom=False]]]]]]]) -> str :\n",
      "            Returns the canonical SMILES string for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - isomericSmiles: (optional) include information about stereochemistry in\n",
      "                  the SMILES.  Defaults to true.\n",
      "                - kekuleSmiles: (optional) use the Kekule form (no aromatic bonds) in\n",
      "                  the SMILES.  Defaults to false.\n",
      "                - rootedAtAtom: (optional) if non-negative, this forces the SMILES \n",
      "                  to start at a particular atom. Defaults to -1.\n",
      "                - canonical: (optional) if false no attempt will be made to canonicalize\n",
      "                  the molecule. Defaults to true.\n",
      "                - allBondsExplicit: (optional) if true, all bond orders will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "                - allHsExplicit: (optional) if true, all H counts will be explicitly indicated\n",
      "                  in the output SMILES. Defaults to false.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToSmiles(RDKit::ROMol [,bool=True [,bool=False [,int=-1 [,bool=True [,bool=False [,bool=False [,bool=False]]]]]]])\n",
      "    \n",
      "    MolToTPLBlock(...)\n",
      "        MolToTPLBlock( (Mol)mol [, (str)partialChargeProp='_GasteigerCharge' [, (bool)writeFirstConfTwice=False]]) -> str :\n",
      "            Returns the Tpl block for a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - partialChargeProp: name of the property to use for partial charges\n",
      "                  Defaults to '_GasteigerCharge'.\n",
      "                - writeFirstConfTwice: Defaults to False.\n",
      "                  This should be set to True when writing TPLs to be read by \n",
      "                  the CombiCode.\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToTPLBlock(RDKit::ROMol [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='_GasteigerCharge' [,bool=False]])\n",
      "    \n",
      "    MolToTPLFile(...)\n",
      "        MolToTPLFile( (Mol)mol, (str)fileName [, (str)partialChargeProp='_GasteigerCharge' [, (bool)writeFirstConfTwice=False]]) -> None :\n",
      "            Writes a molecule to a TPL file.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - fileName: name of the file to write\n",
      "                - partialChargeProp: name of the property to use for partial charges\n",
      "                  Defaults to '_GasteigerCharge'.\n",
      "                - writeFirstConfTwice: Defaults to False.\n",
      "                  This should be set to True when writing TPLs to be read by \n",
      "                  the CombiCode.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToTPLFile(RDKit::ROMol,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='_GasteigerCharge' [,bool=False]])\n",
      "    \n",
      "    MolToXYZBlock(...)\n",
      "        MolToXYZBlock( (Mol)mol [, (int)confId=-1]) -> str :\n",
      "            Returns a XYZ block for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "            \n",
      "              RETURNS:\n",
      "            \n",
      "                a string\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolToXYZBlock(RDKit::ROMol [,int=-1])\n",
      "    \n",
      "    MolToXYZFile(...)\n",
      "        MolToXYZFile( (Mol)mol, (str)filename [, (int)confId=-1]) -> None :\n",
      "            Writes a XYZ file for a molecule\n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule\n",
      "                - filename: the file to write to\n",
      "                - confId: (optional) selects which conformation to output (-1 = default)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void MolToXYZFile(RDKit::ROMol,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,int=-1])\n",
      "    \n",
      "    MolsToJSON(...)\n",
      "        MolsToJSON( (AtomPairsParameters)mols) -> str :\n",
      "            Convert a set of molecules to JSON\n",
      "            \n",
      "                ARGUMENTS:\n",
      "                  - mols: the molecules to work with\n",
      "                RETURNS:\n",
      "                  a string\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > MolsToJSON(boost::python::api::object)\n",
      "    \n",
      "    MurckoDecompose(...)\n",
      "        MurckoDecompose( (Mol)mol) -> Mol :\n",
      "            Do a Murcko decomposition and return the scaffold\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* MurckoDecompose(RDKit::ROMol)\n",
      "    \n",
      "    ParseMolQueryDefFile(...)\n",
      "        ParseMolQueryDefFile( (AtomPairsParameters)fileobj [, (bool)standardize=True [, (str)delimiter='\\t' [, (str)comment='//' [, (int)nameColumn=0 [, (int)smartsColumn=1]]]]]) -> dict :\n",
      "            reads query definitions from a simply formatted file\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::dict ParseMolQueryDefFile(boost::python::api::object {lvalue} [,bool=True [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='\\t' [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >='//' [,unsigned int=0 [,unsigned int=1]]]]])\n",
      "    \n",
      "    PathToSubmol(...)\n",
      "        PathToSubmol( (Mol)mol, (AtomPairsParameters)path [, (bool)useQuery=False [, (AtomPairsParameters)atomMap=None]]) -> Mol :\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* PathToSubmol(RDKit::ROMol,boost::python::api::object {lvalue} [,bool=False [,boost::python::api::object=None]])\n",
      "    \n",
      "    PatternFingerprint(...)\n",
      "        PatternFingerprint( (Mol)mol [, (int)fpSize=2048 [, (list)atomCounts=[] [, (ExplicitBitVect)setOnlyBits=None]]]) -> ExplicitBitVect :\n",
      "            A fingerprint using SMARTS patterns \n",
      "            \n",
      "              NOTE: This function is experimental. The API or results may change from\n",
      "                release to release.\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* PatternFingerprint(RDKit::ROMol [,unsigned int=2048 [,boost::python::list=[] [,ExplicitBitVect*=None]]])\n",
      "    \n",
      "    QuickSmartsMatch(smi, sma, unique=True, display=False)\n",
      "    \n",
      "    RDKFingerprint(...)\n",
      "        RDKFingerprint( (Mol)mol [, (int)minPath=1 [, (int)maxPath=7 [, (int)fpSize=2048 [, (int)nBitsPerHash=2 [, (bool)useHs=True [, (float)tgtDensity=0.0 [, (int)minSize=128 [, (bool)branchedPaths=True [, (bool)useBondOrder=True [, (AtomPairsParameters)atomInvariants=0 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)atomBits=None [, (AtomPairsParameters)bitInfo=None]]]]]]]]]]]]]) -> ExplicitBitVect :\n",
      "            Returns an RDKit topological fingerprint for a molecule\n",
      "            \n",
      "              Explanation of the algorithm below.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "                - minPath: (optional) minimum number of bonds to include in the subgraphs\n",
      "                  Defaults to 1.\n",
      "            \n",
      "                - maxPath: (optional) maximum number of bonds to include in the subgraphs\n",
      "                  Defaults to 7.\n",
      "            \n",
      "                - fpSize: (optional) number of bits in the fingerprint\n",
      "                  Defaults to 2048.\n",
      "            \n",
      "                - nBitsPerHash: (optional) number of bits to set per path\n",
      "                  Defaults to 2.\n",
      "            \n",
      "                - useHs: (optional) include paths involving Hs in the fingerprint if the molecule\n",
      "                  has explicit Hs.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - tgtDensity: (optional) fold the fingerprint until this minimum density has\n",
      "                  been reached\n",
      "                  Defaults to 0.\n",
      "            \n",
      "                - minSize: (optional) the minimum size the fingerprint will be folded to when\n",
      "                  trying to reach tgtDensity\n",
      "                  Defaults to 128.\n",
      "            \n",
      "                - branchedPaths: (optional) if set both branched and unbranched paths will be\n",
      "                  used in the fingerprint.\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - useBondOrder: (optional) if set both bond orders will be used in the path hashes\n",
      "                  Defaults to True.\n",
      "            \n",
      "                - atomInvariants: (optional) a sequence of atom invariants to use in the path hashes\n",
      "                  Defaults to empty.\n",
      "            \n",
      "                - fromAtoms: (optional) a sequence of atom indices. If provided, only paths/subgraphs \n",
      "                  starting from these atoms will be used.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "                - atomBits: (optional) an empty list. If provided, the result will contain a list \n",
      "                  containing the bits each atom sets.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "                - bitInfo: (optional) an empty dict. If provided, the result will contain a dict \n",
      "                  with bits as keys and corresponding bond paths as values.\n",
      "                  Defaults to empty.\n",
      "            \n",
      "              RETURNS: a DataStructs.ExplicitBitVect with _fpSize_ bits\n",
      "            \n",
      "              ALGORITHM:\n",
      "            \n",
      "               This algorithm functions by find all subgraphs between minPath and maxPath in\n",
      "               length.  For each subgraph:\n",
      "            \n",
      "                 1) A hash is calculated.\n",
      "            \n",
      "                 2) The hash is used to seed a random-number generator\n",
      "            \n",
      "                 3) _nBitsPerHash_ random numbers are generated and used to set the corresponding\n",
      "                    bits in the fingerprint\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                ExplicitBitVect* RDKFingerprint(RDKit::ROMol [,unsigned int=1 [,unsigned int=7 [,unsigned int=2048 [,unsigned int=2 [,bool=True [,double=0.0 [,unsigned int=128 [,bool=True [,bool=True [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=None [,boost::python::api::object=None]]]]]]]]]]]]])\n",
      "    \n",
      "    RemoveHs(...)\n",
      "        RemoveHs( (Mol)mol [, (bool)implicitOnly=False [, (bool)updateExplicitCount=False [, (bool)sanitize=True]]]) -> Mol :\n",
      "            Removes any hydrogens from the graph of a molecule.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - implicitOnly: (optional) if this toggle is set, only implicit Hs will\n",
      "                  be removed from the graph.  Default value is 0 (remove implicit and explicit Hs).\n",
      "            \n",
      "                - updateExplicitCount: (optional) if this toggle is set, the explicit H count on atoms with \n",
      "                  Hs will be updated. Default value is 0 (do not update explicit H count).\n",
      "            \n",
      "                - sanitize: (optional) if this toggle is set, the molecule will be sanitized after the Hs\n",
      "                  are removed. Default value is 1 (do sanitize).\n",
      "            \n",
      "              RETURNS: a new molecule with the Hs removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "                - Hydrogens which aren't connected to a heavy atom will not be\n",
      "                  removed.  This prevents molecules like [H][H] from having\n",
      "                  all atoms removed.\n",
      "                - Labelled hydrogen (e.g. atoms with atomic number=1, but isotope > 1),\n",
      "                  will not be removed.\n",
      "                - two coordinate Hs, like the central H in C[H-]C, will not be removed\n",
      "                - Hs connected to dummy atoms will not be removed\n",
      "                - Hs that are part of the definition of double bond Stereochemistry\n",
      "                  will not be removed\n",
      "                - Hs that are not connected to anything else will not be removed\n",
      "            \n",
      "             \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* RemoveHs(RDKit::ROMol [,bool=False [,bool=False [,bool=True]]])\n",
      "    \n",
      "    RemoveStereochemistry(...)\n",
      "        RemoveStereochemistry( (Mol)mol) -> None :\n",
      "            Removes all stereochemistry info from the molecule.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void RemoveStereochemistry(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    RenumberAtoms(...)\n",
      "        RenumberAtoms( (Mol)mol, (AtomPairsParameters)newOrder) -> Mol :\n",
      "            Returns a copy of a molecule with renumbered atoms\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - newOrder: the new ordering the atoms (should be numAtoms long)\n",
      "                  for example: if newOrder is [3,2,0,1], then atom 3 in the original \n",
      "                  molecule will be atom 0 in the new one\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* RenumberAtoms(RDKit::ROMol,boost::python::api::object {lvalue})\n",
      "    \n",
      "    ReplaceCore(...)\n",
      "        ReplaceCore( (Mol)mol, (Mol)core, (AtomPairsParameters)matches [, (bool)replaceDummies=True [, (bool)labelByIndex=False [, (bool)requireDummyMatch=False]]]) -> Mol :\n",
      "            Removes the core of a molecule and labels the sidechains with dummy atoms based on\n",
      "            The matches indices given in the matching vector matches.\n",
      "            Calling:\n",
      "              ReplaceCore(mol,core,mol.GetSubstructMatch(core))\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - coreQuery: the molecule to be used as a substructure query for recognizing the core\n",
      "            \n",
      "                - matches: a matching vector of the type returned by mol.GetSubstructMatch(...)\n",
      "            \n",
      "                - replaceDummies: toggles replacement of atoms that match dummies in the query\n",
      "            \n",
      "                - labelByIndex: toggles labeling the attachment point dummy atoms with \n",
      "                  the index of the core atom they're attached to.\n",
      "            \n",
      "                - requireDummyMatch: if the molecule has side chains that attach at points not\n",
      "                  flagged with a dummy, it will be rejected (None is returned)\n",
      "            \n",
      "              RETURNS: a new molecule with the core removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            EXAMPLES:\n",
      "            \n",
      "                >>> from rdkit.Chem import MolToSmiles, MolFromSmiles, ReplaceCore\n",
      "                >>> mol = MolFromSmiles('C1ONNCC1')\n",
      "                >>> core = MolFromSmiles('NN')\n",
      "            \n",
      "                >>> MolToSmiles(ReplaceCore(mol, core, mol.GetSubstructMatch(core)))\n",
      "                '[1*]OCCC[2*]'\n",
      "            \n",
      "                Since NN is symmetric, we should actually get two matches here if we don't\n",
      "                uniquify the matches.\n",
      "            \n",
      "                >>> [MolToSmiles(ReplaceCore(mol, core, match))\n",
      "                ...     for match in mol.GetSubstructMatches(core, uniquify=False)]\n",
      "                ['[1*]OCCC[2*]', '[1*]CCCO[2*]']\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReplaceCore(RDKit::ROMol,RDKit::ROMol,boost::python::api::object [,bool=True [,bool=False [,bool=False]]])\n",
      "        \n",
      "        ReplaceCore( (Mol)mol, (Mol)coreQuery [, (bool)replaceDummies=True [, (bool)labelByIndex=False [, (bool)requireDummyMatch=False [, (bool)useChirality=False]]]]) -> Mol :\n",
      "            Removes the core of a molecule and labels the sidechains with dummy atoms.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - coreQuery: the molecule to be used as a substructure query for recognizing the core\n",
      "            \n",
      "                - replaceDummies: toggles replacement of atoms that match dummies in the query\n",
      "            \n",
      "                - labelByIndex: toggles labeling the attachment point dummy atoms with \n",
      "                  the index of the core atom they're attached to.\n",
      "            \n",
      "                - requireDummyMatch: if the molecule has side chains that attach at points not\n",
      "                  flagged with a dummy, it will be rejected (None is returned)\n",
      "            \n",
      "                - useChirality: use chirality matching in the coreQuery\n",
      "            \n",
      "              RETURNS: a new molecule with the core removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               >>> from rdkit.Chem import MolToSmiles, MolFromSmiles, MolFromSmarts, ReplaceCore\n",
      "            \n",
      "               Basic usage: remove a core as specified by SMILES (or another molecule).\n",
      "               To get the atom labels which are stored as an isotope of the matched atom, \n",
      "               the output must be written as isomeric smiles.  \n",
      "               A small confusion is that atom isotopes of 0 aren't shown in smiles strings.\n",
      "            \n",
      "               Here we remove a ring and leave the decoration (r-group) behind.\n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('CCCC1CCC1'),MolFromSmiles('C1CCC1')),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCC'\n",
      "            \n",
      "               The isotope label by default is matched by the first connection found. In order to\n",
      "               indicate which atom the decoration is attached in the core query, use labelByIndex=True.\n",
      "               Here the attachment is from the third atom in the smiles string, which is indexed by 3\n",
      "               in the core, like all good computer scientists expect, atoms indices start at 0.\n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('CCN1CCC1'),MolFromSmiles('C1CCN1'),\n",
      "               ...                         labelByIndex=True),\n",
      "               ...   isomericSmiles=True)\n",
      "               '[3*]CC'\n",
      "            \n",
      "               Non-core matches just return None\n",
      "            \n",
      "               >>> ReplaceCore(MolFromSmiles('CCC1CC1'),MolFromSmiles('C1CCC1'))\n",
      "            \n",
      "               The bond between atoms are considered part of the core and are removed as well\n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CC2C1CCC2'),MolFromSmiles('C1CCC1')),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCC[2*]'\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CNCC1'),MolFromSmiles('N')),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCCC[2*]'\n",
      "            \n",
      "               When using dummy atoms, cores should be read in as SMARTS.  When read as SMILES\n",
      "               dummy atoms only match other dummy atoms.\n",
      "               The replaceDummies flag indicates whether matches to the dummy atoms should be considered as part\n",
      "               of the core or as part of the decoration (r-group)\n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CNCC1'),MolFromSmarts('[*]N[*]'),\n",
      "               ...                         replaceDummies=True),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CC[2*]'\n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CNCC1'),MolFromSmarts('[*]N[*]'),\n",
      "               ...                         replaceDummies=False),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CCCC[2*]'\n",
      "            \n",
      "            \n",
      "               >>> MolToSmiles(ReplaceCore(MolFromSmiles('C1CCC1CN'),MolFromSmarts('C1CCC1[*]'),\n",
      "               ...                         replaceDummies=False),\n",
      "               ...             isomericSmiles=True)\n",
      "               '[1*]CN'\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReplaceCore(RDKit::ROMol,RDKit::ROMol [,bool=True [,bool=False [,bool=False [,bool=False]]]])\n",
      "    \n",
      "    ReplaceSidechains(...)\n",
      "        ReplaceSidechains( (Mol)mol, (Mol)coreQuery [, (bool)useChirality=False]) -> Mol :\n",
      "            Replaces sidechains in a molecule with dummy atoms for their attachment points.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - coreQuery: the molecule to be used as a substructure query for recognizing the core\n",
      "            \n",
      "                - useChirality: (optional) match the substructure query using chirality\n",
      "            \n",
      "              RETURNS: a new molecule with the sidechains removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               The following examples substitute SMILES/SMARTS strings for molecules, you'd have\n",
      "               to actually use molecules:\n",
      "            \n",
      "                - ReplaceSidechains('CCC1CCC1','C1CCC1') -> '[Xa]C1CCC1'\n",
      "            \n",
      "                - ReplaceSidechains('CCC1CC1','C1CCC1') -> ''\n",
      "            \n",
      "                - ReplaceSidechains('C1CC2C1CCC2','C1CCC1') -> '[Xa]C1CCC1[Xb]'\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::ROMol* ReplaceSidechains(RDKit::ROMol,RDKit::ROMol [,bool=False])\n",
      "    \n",
      "    ReplaceSubstructs(...)\n",
      "        ReplaceSubstructs( (Mol)mol, (Mol)query, (Mol)replacement [, (bool)replaceAll=False [, (int)replacementConnectionPoint=0 [, (bool)useChirality=False]]]) -> object :\n",
      "            Replaces atoms matching a substructure query in a molecule\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "                - query: the molecule to be used as a substructure query\n",
      "            \n",
      "                - replacement: the molecule to be used as the replacement\n",
      "            \n",
      "                - replaceAll: (optional) if this toggle is set, all substructures matching\n",
      "                  the query will be replaced in a single result, otherwise each result will\n",
      "                  contain a separate replacement.\n",
      "                  Default value is False (return multiple replacements)\n",
      "                - replacementConnectionPoint: (optional) index of the atom in the replacement that\n",
      "                  the bond should be made to.\n",
      "                - useChirality: (optional) match the substructure query using chirality\n",
      "            \n",
      "              RETURNS: a tuple of new molecules with the substructures replaced removed\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The original molecule is *not* modified.\n",
      "            \n",
      "              EXAMPLES:\n",
      "            \n",
      "               The following examples substitute SMILES/SMARTS strings for molecules, you'd have\n",
      "               to actually use molecules:\n",
      "            \n",
      "                - ReplaceSubstructs('CCOC','OC','NC') -> ('CCNC',)\n",
      "            \n",
      "                - ReplaceSubstructs('COCCOC','OC','NC') -> ('COCCNC','CNCCOC')\n",
      "            \n",
      "                - ReplaceSubstructs('COCCOC','OC','NC',True) -> ('CNCCNC',)\n",
      "            \n",
      "                - ReplaceSubstructs('COCCOC','OC','CN',True,1) -> ('CNCCNC',)\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                _object* ReplaceSubstructs(RDKit::ROMol,RDKit::ROMol,RDKit::ROMol [,bool=False [,unsigned int=0 [,bool=False]]])\n",
      "    \n",
      "    SanitizeMol(...)\n",
      "        SanitizeMol( (Mol)mol [, (int)sanitizeOps=rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL [, (bool)catchErrors=False]]) -> SanitizeFlags :\n",
      "            Kekulize, check valencies, set aromaticity, conjugation and hybridization\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "                - If sanitization fails, an exception will be thrown unless catchErrors is set\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to be modified\n",
      "                - sanitizeOps: (optional) sanitization operations to be carried out\n",
      "                  these should be constructed by or'ing together the\n",
      "                  operations in rdkit.Chem.SanitizeFlags\n",
      "                - catchErrors: (optional) if provided, instead of raising an exception\n",
      "                  when sanitization fails (the default behavior), the \n",
      "                  first operation that failed (as defined in rdkit.Chem.SanitizeFlags)\n",
      "                  is returned. Zero is returned on success.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::MolOps::SanitizeFlags SanitizeMol(RDKit::ROMol {lvalue} [,unsigned long long=rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL [,bool=False]])\n",
      "    \n",
      "    SetAromaticity(...)\n",
      "        SetAromaticity( (Mol)mol [, (AromaticityModel)model=rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_DEFAULT]) -> None :\n",
      "            does aromaticity perception\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - model: the model to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAromaticity(RDKit::ROMol {lvalue} [,RDKit::MolOps::AromaticityModel=rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_DEFAULT])\n",
      "    \n",
      "    SetAtomAlias(...)\n",
      "        SetAtomAlias( (Atom)atom, (str)rlabel) -> None :\n",
      "            Sets the atom's MDL alias text.\n",
      "            Setting to an empty string clears the alias.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAtomAlias(RDKit::Atom*,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    SetAtomRLabel(...)\n",
      "        SetAtomRLabel( (Atom)atom, (int)rlabel) -> None :\n",
      "            Sets the atom's MDL RLabel (this is an integer from 0 to 99).\n",
      "            Setting to 0 clears the rlabel.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAtomRLabel(RDKit::Atom*,int)\n",
      "    \n",
      "    SetAtomValue(...)\n",
      "        SetAtomValue( (Atom)atom, (str)rlabel) -> None :\n",
      "            Sets the atom's MDL alias text.\n",
      "            Setting to an empty string clears the alias.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetAtomValue(RDKit::Atom*,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    SetBondStereoFromDirections(...)\n",
      "        SetBondStereoFromDirections( (Mol)mol) -> None :\n",
      "            Uses the directions of neighboring bonds to set cis/trans stereo on double bonds.\n",
      "                    \n",
      "              ARGUMENTS:\n",
      "              \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetBondStereoFromDirections(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    SetConjugation(...)\n",
      "        SetConjugation( (Mol)mol) -> None :\n",
      "            finds conjugated bonds\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetConjugation(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    SetDefaultPickleProperties(...)\n",
      "        SetDefaultPickleProperties( (int)arg1) -> None :\n",
      "            Set the current global mol pickler options.\n",
      "        \n",
      "            C++ signature :\n",
      "                void SetDefaultPickleProperties(unsigned int)\n",
      "    \n",
      "    SetDoubleBondNeighborDirections(...)\n",
      "        SetDoubleBondNeighborDirections( (Mol)mol [, (AtomPairsParameters)conf=None]) -> None :\n",
      "            Uses the stereo info on double bonds to set the directions of neighboring single bonds\n",
      "                    \n",
      "              ARGUMENTS:\n",
      "              \n",
      "                - mol: the molecule to be modified\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetDoubleBondNeighborDirections(RDKit::ROMol {lvalue} [,boost::python::api::object=None])\n",
      "    \n",
      "    SetHybridization(...)\n",
      "        SetHybridization( (Mol)mol) -> None :\n",
      "            Assigns hybridization states to atoms\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "            \n",
      "              NOTES:\n",
      "            \n",
      "                - The molecule is modified in place.\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetHybridization(RDKit::ROMol {lvalue})\n",
      "    \n",
      "    SetSupplementalSmilesLabel(...)\n",
      "        SetSupplementalSmilesLabel( (Atom)atom, (str)label) -> None :\n",
      "            Sets a supplemental label on an atom that is written to the smiles string.\n",
      "            \n",
      "            >>> m = Chem.MolFromSmiles(\"C\")\n",
      "            >>> Chem.SetSupplementalSmilesLabel(m.GetAtomWithIdx(0), '<xxx>')\n",
      "            >>> Chem.MolToSmiles(m)\n",
      "            'C<xxx>'\n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void SetSupplementalSmilesLabel(RDKit::Atom*,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)\n",
      "    \n",
      "    SmilesMolSupplierFromText(...)\n",
      "        SmilesMolSupplierFromText( (str)text [, (str)delimiter=' ' [, (int)smilesColumn=0 [, (int)nameColumn=1 [, (bool)titleLine=True [, (bool)sanitize=True]]]]]) -> SmilesMolSupplier :\n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SmilesMolSupplier* SmilesMolSupplierFromText(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > [,std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >=' ' [,int=0 [,int=1 [,bool=True [,bool=True]]]]])\n",
      "    \n",
      "    SplitMolByPDBChainId(...)\n",
      "        SplitMolByPDBChainId( (Mol)mol [, (AtomPairsParameters)whiteList=None [, (bool)negateList=False]]) -> dict :\n",
      "            Splits a molecule into pieces based on PDB chain information.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - whiteList: only residues in this list will be returned\n",
      "                - negateList: if set, negates the white list inclusion logic\n",
      "            \n",
      "              RETURNS: a dictionary keyed by chain id with molecules as the values\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::dict SplitMolByPDBChainId(RDKit::ROMol [,boost::python::api::object=None [,bool=False]])\n",
      "    \n",
      "    SplitMolByPDBResidues(...)\n",
      "        SplitMolByPDBResidues( (Mol)mol [, (AtomPairsParameters)whiteList=None [, (bool)negateList=False]]) -> dict :\n",
      "            Splits a molecule into pieces based on PDB residue information.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - mol: the molecule to use\n",
      "                - whiteList: only residues in this list will be returned\n",
      "                - negateList: if set, negates the white list inclusion logic\n",
      "            \n",
      "              RETURNS: a dictionary keyed by residue name with molecules as the values\n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                boost::python::dict SplitMolByPDBResidues(RDKit::ROMol [,boost::python::api::object=None [,bool=False]])\n",
      "    \n",
      "    SupplierFromFilename(fileN, delim='', **kwargs)\n",
      "    \n",
      "    UnfoldedRDKFingerprintCountBased(...)\n",
      "        UnfoldedRDKFingerprintCountBased( (Mol)mol [, (int)minPath=1 [, (int)maxPath=7 [, (bool)useHs=True [, (bool)branchedPaths=True [, (bool)useBondOrder=True [, (AtomPairsParameters)atomInvariants=0 [, (AtomPairsParameters)fromAtoms=0 [, (AtomPairsParameters)atomBits=None [, (AtomPairsParameters)bitInfo=None]]]]]]]]]) -> ULongSparseIntVect :\n",
      "            Returns an unfolded count-based version of the RDKit fingerprint for a molecule\n",
      "            \n",
      "            ARGUMENTS:\n",
      "                \n",
      "                    - mol: the molecule to use\n",
      "                \n",
      "                    - minPath: (optional) minimum number of bonds to include in the subgraphs\n",
      "                      Defaults to 1.\n",
      "                \n",
      "                    - maxPath: (optional) maximum number of bonds to include in the subgraphs\n",
      "                      Defaults to 7.\n",
      "                \n",
      "                    - useHs: (optional) include paths involving Hs in the fingerprint if the molecule\n",
      "                      has explicit Hs.\n",
      "                      Defaults to True.\n",
      "                \n",
      "                    - branchedPaths: (optional) if set both branched and unbranched paths will be\n",
      "                      used in the fingerprint.\n",
      "                      Defaults to True.\n",
      "                \n",
      "                    - useBondOrder: (optional) if set both bond orders will be used in the path hashes\n",
      "                      Defaults to True.\n",
      "                \n",
      "                    - atomInvariants: (optional) a sequence of atom invariants to use in the path hashes\n",
      "                      Defaults to empty.\n",
      "                \n",
      "                    - fromAtoms: (optional) a sequence of atom indices. If provided, only paths/subgraphs \n",
      "                      starting from these atoms will be used.\n",
      "                      Defaults to empty.\n",
      "                \n",
      "                    - atomBits: (optional) an empty list. If provided, the result will contain a list \n",
      "                      containing the bits each atom sets.\n",
      "                      Defaults to empty.\n",
      "                \n",
      "                    - bitInfo: (optional) an empty dict. If provided, the result will contain a dict \n",
      "                      with bits as keys and corresponding bond paths as values.\n",
      "                      Defaults to empty.\n",
      "                 \n",
      "                 \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                RDKit::SparseIntVect<unsigned long long>* UnfoldedRDKFingerprintCountBased(RDKit::ROMol [,unsigned int=1 [,unsigned int=7 [,bool=True [,bool=True [,bool=True [,boost::python::api::object=0 [,boost::python::api::object=0 [,boost::python::api::object=None [,boost::python::api::object=None]]]]]]]]])\n",
      "    \n",
      "    WedgeBond(...)\n",
      "        WedgeBond( (Bond)arg1, (int)arg2, (Conformer)arg3) -> None :\n",
      "            Set the wedging on an individual bond from a molecule.\n",
      "               The wedging scheme used is that from Mol files.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - bond: the bond to update\n",
      "                - atom ID: the atom from which to do the wedging\n",
      "                - conformer: the conformer to use to determine wedge direction\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void WedgeBond(RDKit::Bond*,unsigned int,RDKit::Conformer const*)\n",
      "    \n",
      "    WedgeMolBonds(...)\n",
      "        WedgeMolBonds( (Mol)arg1, (Conformer)arg2) -> None :\n",
      "            Set the wedging on single bonds in a molecule.\n",
      "               The wedging scheme used is that from Mol files.\n",
      "            \n",
      "              ARGUMENTS:\n",
      "            \n",
      "                - molecule: the molecule to update\n",
      "                - conformer: the conformer to use to determine wedge direction\n",
      "            \n",
      "            \n",
      "            \n",
      "        \n",
      "            C++ signature :\n",
      "                void WedgeMolBonds(RDKit::ROMol {lvalue},RDKit::Conformer const*)\n",
      "    \n",
      "    WrapLogs(...)\n",
      "        WrapLogs() -> None :\n",
      "            Wrap the internal RDKit streams so they go to python's SysStdErr\n",
      "        \n",
      "            C++ signature :\n",
      "                void WrapLogs()\n",
      "    \n",
      "    tossit(...)\n",
      "        tossit() -> None :\n",
      "        \n",
      "            C++ signature :\n",
      "                void tossit()\n",
      "\n",
      "DATA\n",
      "    ADJUST_IGNOREALL = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST_IG...\n",
      "    ADJUST_IGNORECHAINS = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST...\n",
      "    ADJUST_IGNOREDUMMIES = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUS...\n",
      "    ADJUST_IGNORENONDUMMIES = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.AD...\n",
      "    ADJUST_IGNORENONE = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST_I...\n",
      "    ADJUST_IGNORERINGS = rdkit.Chem.rdmolops.AdjustQueryWhichFlags.ADJUST_...\n",
      "    ALLOW_CHARGE_SEPARATION = rdkit.Chem.rdchem.ResonanceFlags.ALLOW_CHARG...\n",
      "    ALLOW_INCOMPLETE_OCTETS = rdkit.Chem.rdchem.ResonanceFlags.ALLOW_INCOM...\n",
      "    AROMATICITY_CUSTOM = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_...\n",
      "    AROMATICITY_DEFAULT = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY...\n",
      "    AROMATICITY_MDL = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_MDL\n",
      "    AROMATICITY_RDKIT = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_R...\n",
      "    AROMATICITY_SIMPLE = rdkit.Chem.rdmolops.AromaticityModel.AROMATICITY_...\n",
      "    AllProps = rdkit.Chem.rdchem.PropertyPickleOptions.AllProps\n",
      "    AtomProps = rdkit.Chem.rdchem.PropertyPickleOptions.AtomProps\n",
      "    BondProps = rdkit.Chem.rdchem.PropertyPickleOptions.BondProps\n",
      "    CHI_OTHER = rdkit.Chem.rdchem.ChiralType.CHI_OTHER\n",
      "    CHI_TETRAHEDRAL_CCW = rdkit.Chem.rdchem.ChiralType.CHI_TETRAHEDRAL_CCW\n",
      "    CHI_TETRAHEDRAL_CW = rdkit.Chem.rdchem.ChiralType.CHI_TETRAHEDRAL_CW\n",
      "    CHI_UNSPECIFIED = rdkit.Chem.rdchem.ChiralType.CHI_UNSPECIFIED\n",
      "    COMPOSITE_AND = rdkit.Chem.rdchem.CompositeQueryType.COMPOSITE_AND\n",
      "    COMPOSITE_OR = rdkit.Chem.rdchem.CompositeQueryType.COMPOSITE_OR\n",
      "    COMPOSITE_XOR = rdkit.Chem.rdchem.CompositeQueryType.COMPOSITE_XOR\n",
      "    ComputedProps = rdkit.Chem.rdchem.PropertyPickleOptions.ComputedProps\n",
      "    INCHI_AVAILABLE = True\n",
      "    KEKULE_ALL = rdkit.Chem.rdchem.ResonanceFlags.KEKULE_ALL\n",
      "    LayeredFingerprint_substructLayers = 7\n",
      "    MolProps = rdkit.Chem.rdchem.PropertyPickleOptions.MolProps\n",
      "    NoProps = rdkit.Chem.rdchem.PropertyPickleOptions.NoProps\n",
      "    PrivateProps = rdkit.Chem.rdchem.PropertyPickleOptions.PrivateProps\n",
      "    QueryAtomData = rdkit.Chem.rdchem.PropertyPickleOptions.QueryAtomData\n",
      "    SANITIZE_ADJUSTHS = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ADJUSTH...\n",
      "    SANITIZE_ALL = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_ALL\n",
      "    SANITIZE_CLEANUP = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_CLEANUP\n",
      "    SANITIZE_CLEANUPCHIRALITY = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE...\n",
      "    SANITIZE_FINDRADICALS = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_FIN...\n",
      "    SANITIZE_KEKULIZE = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_KEKULIZ...\n",
      "    SANITIZE_NONE = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE\n",
      "    SANITIZE_PROPERTIES = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_PROPE...\n",
      "    SANITIZE_SETAROMATICITY = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_S...\n",
      "    SANITIZE_SETCONJUGATION = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_S...\n",
      "    SANITIZE_SETHYBRIDIZATION = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE...\n",
      "    SANITIZE_SYMMRINGS = rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_SYMMRI...\n",
      "    STEREO_ABSOLUTE = rdkit.Chem.rdchem.StereoGroupType.STEREO_ABSOLUTE\n",
      "    STEREO_AND = rdkit.Chem.rdchem.StereoGroupType.STEREO_AND\n",
      "    STEREO_OR = rdkit.Chem.rdchem.StereoGroupType.STEREO_OR\n",
      "    UNCONSTRAINED_ANIONS = rdkit.Chem.rdchem.ResonanceFlags.UNCONSTRAINED_...\n",
      "    UNCONSTRAINED_CATIONS = rdkit.Chem.rdchem.ResonanceFlags.UNCONSTRAINED...\n",
      "    templDir = '/opt/anaconda3/envs/moldist2/share/RDKit/Data/'\n",
      "\n",
      "FILE\n",
      "    /opt/anaconda3/envs/moldist2/lib/python3.7/site-packages/rdkit/Chem/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(rd.Chem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppl = rd.Chem.SDMolSupplier('../data/raw/ChemDivFull.sdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<rdkit.Chem.rdmolfiles.SDMolSupplier object at 0xa1d9a4c30>\n"
     ]
    }
   ],
   "source": [
    "print(suppl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12 ['CDIV0001', 'CDIV0011', 'CDIV0021', 'CDIV0031', 'CDIV0041', 'CDIV0051', 'CDIV0061', 'CDIV0071', 'CDIV0081', 'CDIV0091', 'CDIV0101', 'CDIV0111']\n"
     ]
    }
   ],
   "source": [
    "# Extract only the 960 tested molecules\n",
    "plates = [\"CDIV%04d\"%p for p in range(1,121,10)]\n",
    "print(len(plates),plates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50000 960\n"
     ]
    }
   ],
   "source": [
    "tested = [x for x in suppl if x.GetProp(\"BATCH_PLATE\") in plates]\n",
    "print(len(suppl),len(tested))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11  CDIV0001 A02\n",
      "24  CDIV0011 A02\n",
      "26  CDIV0021 A02\n",
      "23  CDIV0031 A02\n",
      "23  CDIV0041 A02\n",
      "28  CDIV0051 A02\n",
      "24  CDIV0061 A02\n",
      "20  CDIV0071 A02\n",
      "24  CDIV0081 A02\n",
      "23  CDIV0091 A02\n",
      "29  CDIV0101 A02\n",
      "20  CDIV0111 A02\n"
     ]
    }
   ],
   "source": [
    "for mol in tested:\n",
    "    if mol is None: continue\n",
    "    if mol.GetProp(\"BATCH_WELL\")==\"A02\":\n",
    "        print(mol.GetNumAtoms(),mol.GetProp(\"_Name\"),mol.GetProp(\"BATCH_PLATE\"),mol.GetProp(\"BATCH_WELL\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "__computedProps [numArom,_StereochemDone,]\n",
      "_Name \n",
      "_MolFileInfo   -ISIS-  01311913252D\n",
      "_MolFileComments \n",
      "numArom 3\n",
      "_StereochemDone 1\n",
      "compound_Corp_Reg_Number STF-085941\n",
      "rotatable_bonds 6.000000000000000e+000\n",
      "BATCH_NUMBER 1\n",
      "BATCH_MW 3.824600000000000e+002\n",
      "BATCH_VENDOR_CATALOG 6024-0185\n",
      "BATCH_PLATE CDIV0111\n",
      "BATCH_WELL H11\n",
      "LogP 5.249700000000000e+000\n",
      "LogS -6.007000000000000e+000\n",
      "H_Acceptors 4\n",
      "H_Donors 1\n",
      "AnlTst_Display NMR\n",
      "BatAnl_Link D:\\Spectra\\FirstSide\\10K\\6024-0185.emf\n",
      "VENDOR_DISPLAY ChemDiv\n"
     ]
    }
   ],
   "source": [
    "for x in mol.GetPropNames(includePrivate=True,includeComputed=True):\n",
    "    print(x,mol.GetProp(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllChem.Compute2DCoords(mol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<rdkit.Chem.rdchem.Mol object at 0xa1d9d4a30>\n"
     ]
    }
   ],
   "source": [
    "print(mol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "m2=rd.Chem.AddHs(mol)\n",
    "AllChem.EmbedMolecule(m2)\n",
    "AllChem.MMFFOptimizeMolecule(m2)\n",
    "m2=rd.Chem.RemoveHs(m2)\n",
    "img=Draw.MolsToGridImage([mol,m2],molsPerRow=2,subImgSize=(200,200),legends=[\"2d\",\"3d\"])\n",
    "img.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     RDKit          2D\n",
      "\n",
      " 20 22  0  0  0  0  0  0  0  0999 V2000\n",
      "    0.5917   -1.2875    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    0.0125   -0.4667    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    1.5917   -1.2875    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    0.0125   -2.0792    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    2.1167   -0.4042    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "   -0.9250   -0.7750    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    2.1167   -2.1375    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "   -0.9250   -1.7750    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    2.6250   -3.0167    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    1.5917    0.4458    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    0.5542    2.1417    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    0.0417    3.0250    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    0.5917    0.4083    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    2.1167    1.2958    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    0.1042    1.2958    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "    1.5667    2.1833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "   -1.7833   -0.2875    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "   -1.7833   -2.2917    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "   -2.6333   -1.7750    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "   -2.6333   -0.7750    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n",
      "  2  1  2  0\n",
      "  3  1  1  0\n",
      "  4  1  1  0\n",
      "  5  3  2  0\n",
      "  6  2  1  0\n",
      "  7  3  1  0\n",
      "  8  4  1  0\n",
      "  9  7  3  0\n",
      " 10  5  1  0\n",
      " 11 16  2  0\n",
      " 12 11  1  0\n",
      " 13 10  1  0\n",
      " 14 10  2  0\n",
      " 15 13  2  0\n",
      " 16 14  1  0\n",
      " 17  6  2  0\n",
      " 18  8  2  0\n",
      " 19 18  1  0\n",
      " 20 17  1  0\n",
      "  8  6  1  0\n",
      " 20 19  2  0\n",
      " 11 15  1  0\n",
      "M  END\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(rd.Chem.MolToMolBlock(tested[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getMol(mols,plate,well):\n",
    "    mol=[x for x in mols if x.GetProp(\"BATCH_PLATE\")==\"CDIV%04d\"%plate and x.GetProp(\"BATCH_WELL\")==well]\n",
    "    assert(len(mol)==1)\n",
    "    return mol[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare some molecules which give the same response\n",
    "m91d2=getMol(tested,91,'D02')\n",
    "m91e2=getMol(tested,91,\"E02\")\n",
    "img=Draw.MolsToGridImage([m91d2,m91e2],molsPerRow=2,subImgSize=(200,200),legends=[\"91E2\",\"91D2\"])\n",
    "img.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[#6]12:[#6](:[#7]:[#6]3:[#7](:[#6]:1=[#8]):[#6]:[#6]:[#6]:[#6]:3):[#7](:[#6](:[#6](:[#6]:2)-[#6](=[#8])-[#7]-[#6])=[#7])-[#6]-[#6]-[#6]-[#8]-[#6]'"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from rdkit.Chem import rdFMCS\n",
    "res=rdFMCS.FindMCS([m91d2,m91e2])\n",
    "res.smartsString"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<rdkit.Chem.rdFMCS.MCSResult at 0xa22abfd00>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check chemical features\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import ChemicalFeatures\n",
    "from rdkit import RDConfig\n",
    "import os\n",
    "fdefName = os.path.join(RDConfig.RDDataDir,'BaseFeatures.fdef')\n",
    "factory = ChemicalFeatures.BuildFeatureFactory(fdefName)\n",
    "feats = [factory.GetFeaturesForMol(x) for x in [m91d2,m91e2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(feats[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on MolChemicalFeature in module rdkit.Chem.rdMolChemicalFeatures object:\n",
      "\n",
      "class MolChemicalFeature(Boost.Python.instance)\n",
      " |  Class to represent a chemical feature.\n",
      " |  These chemical features may or may not have been derived from molecule object;\n",
      " |  i.e. it is possible to have a chemical feature that was created just from its type\n",
      " |  and location.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      MolChemicalFeature\n",
      " |      Boost.Python.instance\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  ClearCache(...)\n",
      " |      ClearCache( (MolChemicalFeature)arg1) -> None :\n",
      " |          Clears the cache used to store position information.\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              void ClearCache(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  GetActiveConformer(...)\n",
      " |      GetActiveConformer( (MolChemicalFeature)arg1) -> int :\n",
      " |          Gets the conformer to use.\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              int GetActiveConformer(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  GetAtomIds(...)\n",
      " |      GetAtomIds( (MolChemicalFeature)arg1) -> object :\n",
      " |          Get the IDs of the atoms that participate in the feature\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              _object* GetAtomIds(RDKit::MolChemicalFeature)\n",
      " |  \n",
      " |  GetFactory(...)\n",
      " |      GetFactory( (MolChemicalFeature)arg1) -> MolChemicalFeatureFactory :\n",
      " |          Get the factory used to generate this feature\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              RDKit::MolChemicalFeatureFactory const* GetFactory(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  GetFamily(...)\n",
      " |      GetFamily( (MolChemicalFeature)arg1) -> str :\n",
      " |          Get the family to which the feature belongs; donor, acceptor, etc.\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > GetFamily(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  GetId(...)\n",
      " |      GetId( (MolChemicalFeature)arg1) -> int :\n",
      " |          Returns the identifier of the feature\n",
      " |          \n",
      " |      \n",
      " |          C++ signature :\n",
      " |              int GetId(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  GetMol(...)\n",
      " |      GetMol( (MolChemicalFeature)arg1) -> Mol :\n",
      " |          Get the molecule used to derive the features\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              RDKit::ROMol const* GetMol(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  GetPos(...)\n",
      " |      GetPos( (MolChemicalFeature)self [, (int)confId=-1]) -> Point3D :\n",
      " |          Get the location of the chemical feature\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              RDGeom::Point3D GetPos(RDKit::MolChemicalFeature {lvalue} [,int=-1])\n",
      " |  \n",
      " |  GetType(...)\n",
      " |      GetType( (MolChemicalFeature)arg1) -> str :\n",
      " |          Get the specific type for the feature\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > GetType(RDKit::MolChemicalFeature {lvalue})\n",
      " |  \n",
      " |  SetActiveConformer(...)\n",
      " |      SetActiveConformer( (MolChemicalFeature)arg1, (int)arg2) -> None :\n",
      " |          Sets the conformer to use (must be associated with a molecule).\n",
      " |      \n",
      " |          C++ signature :\n",
      " |              void SetActiveConformer(RDKit::MolChemicalFeature {lvalue},int)\n",
      " |  \n",
      " |  __init__(...)\n",
      " |      Raises an exception\n",
      " |      This class cannot be instantiated from Python\n",
      " |  \n",
      " |  __reduce__ = <unnamed Boost.Python function>(...)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods inherited from Boost.Python.instance:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from Boost.Python.class\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Boost.Python.instance:\n",
      " |  \n",
      " |  __dict__\n",
      " |  \n",
      " |  __weakref__\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(feats[0][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Donor SingleAtomDonor (11,)\n",
      "Donor SingleAtomDonor (19,)\n",
      "Acceptor SingleAtomAcceptor (2,)\n",
      "Acceptor SingleAtomAcceptor (13,)\n",
      "Acceptor SingleAtomAcceptor (15,)\n",
      "Acceptor SingleAtomAcceptor (21,)\n",
      "Aromatic Arom6 (0, 1, 5, 7, 8, 2)\n",
      "Aromatic Arom6 (0, 1, 9, 4, 6, 3)\n",
      "Aromatic Arom6 (7, 8, 14, 17, 16, 12)\n",
      "Hydrophobe ThreeWayAttach (1,)\n",
      "Hydrophobe ThreeWayAttach (4,)\n",
      "Hydrophobe ChainTwoWayAttach (20,)\n",
      "\n",
      "Donor SingleAtomDonor (13,)\n",
      "Donor SingleAtomDonor (19,)\n",
      "Acceptor SingleAtomAcceptor (2,)\n",
      "Acceptor SingleAtomAcceptor (16,)\n",
      "Acceptor SingleAtomAcceptor (17,)\n",
      "Acceptor SingleAtomAcceptor (21,)\n",
      "Aromatic Arom6 (0, 1, 6, 3, 8, 2)\n",
      "Aromatic Arom6 (0, 1, 9, 5, 7, 4)\n",
      "Aromatic Arom6 (3, 8, 12, 15, 14, 10)\n",
      "Hydrophobe ThreeWayAttach (1,)\n",
      "Hydrophobe ThreeWayAttach (5,)\n",
      "Hydrophobe ThreeWayAttach (14,)\n",
      "Hydrophobe ChainTwoWayAttach (20,)\n"
     ]
    }
   ],
   "source": [
    "for fs in feats:\n",
    "    print(\"\")\n",
    "    for f in fs:\n",
    "        print(f.GetFamily(),f.GetType(),f.GetAtomIds())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "File: data/MinimalFeatures.fdef could not be opened.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-62-f91df1c8672a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mrdkit\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mChem\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mChemicalFeatures\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mfdefName\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'data/MinimalFeatures.fdef'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mfeatFactory\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mChemicalFeatures\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mBuildFeatureFactory\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfdefName\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m: File: data/MinimalFeatures.fdef could not be opened."
     ]
    }
   ],
   "source": [
    "# Generate pharmacophore fingerprints \n",
    "from rdkit import Chem\n",
    "from rdkit.Chem import ChemicalFeatures\n",
    "fdefName = 'data/MinimalFeatures.fdef'\n",
    "featFactory = ChemicalFeatures.BuildFeatureFactory(fdefName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdkit.Chem.Pharm2D.SigFactory import SigFactory\n",
    "sigFactory = SigFactory(featFactory,minPointCount=2,maxPointCount=3)\n",
    "sigFactory.SetBins([(0,2),(2,5),(5,8)])\n",
    "sigFactory.Init()\n",
    "sigFactory.GetSigSize()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
